In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Data Cleaning

Przed uruchomieniem naszych danych przez model ważnym krokiem jest **Eksploracyjna analiza danych** lub EDA. EDA, jak sama nazwa wskazuje, to dogłębna analiza naszych danych. Proces EDA przeplata czyszczenie danych, radzenie sobie z brakującymi wartościami oraz wizualizację danych i ich właściwości statystycznych. Zazwyczaj wszystkie te procesy są wykonywane razem, ale dla potrzeb edukacyjnych rozbijemy je na osobne części.

### Prerequisites
[flights.txt](https://drive.google.com/file/d/1cVV3TZcxS31fk9JrskaRP1pbOfaoNcwe/view?usp=sharing)
(źródło: https://www.kaggle.com/mmetter/flights/data).
W większości przypadków dane posiadają dokumentację. **Czytanie dokumentacji danych jest ważne**!!!! Ta konkretna część danych nie ma jednak żadnej dokumentacji - będziemy musieli skorzystać z naszej intuicji oraz nazw kolumn (zmiennych).

### Typy danych

<table >
	<tbody>
		<tr>
            <td><b>Typ danych</b></td>
            <td><b>Typ danych w Pythonie</b></td>
            <td><b>Przykłady</b></td>
		</tr>
		<tr>
			<td>Dane tekstowe</td>
            <td>str</td>
			<td>Nazwiska, adresy</td>
		</tr>
		<tr>
			<td>Integers</td>
            <td>int</td>
			<td># przedmioty, # osoby</td>
		</tr>
		<tr>
			<td>Floats/Ułamki</td>
            <td>float</td>
			<td>Waluty, odległości</td>
		</tr>
		<tr>
			<td>Binary/Boolean</td>
            <td>bool</td>
			<td>Pytania zamknięte, yes/no</td>
		</tr>
		<tr>
			<td>Data (i czas)</td>
            <td>datetime</td>
			<td>Data wysyłki, czas przyjazdu</td>
		</tr>
		<tr>
			<td>Kategorie</td>
            <td>category</td>
			<td>Stany, kolory, płeć</td>
		</tr>
	</tbody>
</table>

In [ ]:
path='sciezka na twoim dysku/komputerze'

In [ ]:
import pandas as pd
pd.set_option('display.max_columns', None)
flights_df = pd.read_csv("flights.txt", sep="|")

In [ ]:
flights_df.head()

TRANSACTIONID  FLIGHTDATE AIRLINECODE                     AIRLINENAME  \
0       54548800    20020101          WN      Southwest Airlines Co.: WN   
1       55872300    20020101          CO  Continental Air Lines Inc.: CO   
2       54388800    20020101          WN      Southwest Airlines Co.: WN   
3       54486500    20020101          WN      Southwest Airlines Co.: WN   
4       55878700    20020103          CO  Continental Air Lines Inc.: CO   

  TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
0  N103@@       1425               ABQ   
1  N83872        150               ABQ   
2  N334@@        249               ABQ   
3  N699@@        902               ABQ   
4  N58606        234               ABQ   

                                    ORIGAIRPORTNAME ORIGINCITYNAME  \
0  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   
1  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   
2  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   
3  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   
4  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   

  ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
0          NM      New Mexico             DAL   
1          NM      New Mexico             IAH   
2          NM      New Mexico             MCI   
3          NM      New Mexico             LAS   
4          NM      New Mexico             IAH   

                                   DESTAIRPORTNAME DESTCITYNAME DESTSTATE  \
0                      DallasTX: Dallas Love Field       Dallas        TX   
1  HoustonTX: George Bush Intercontinental/Houston      Houston        TX   
2         Kansas CityMO: Kansas City International  Kansas City        MO   
3              Las VegasNV: McCarran International    Las Vegas        NV   
4  HoustonTX: George Bush Intercontinental/Houston      Houston        TX   

  DESTSTATENAME  CRSDEPTIME  DEPTIME  DEPDELAY  TAXIOUT  WHEELSOFF  WHEELSON  \
0         Texas        1425   1425.0       0.0      8.0     1433.0    1648.0   
1         Texas        1130   1136.0       6.0     12.0     1148.0    1419.0   
2      Missouri        1215   1338.0      83.0      7.0     1345.0    1618.0   
3        Nevada        1925   1925.0       0.0      5.0     1930.0    1947.0   
4         Texas        1455   1453.0      -2.0     11.0     1504.0    1742.0   

   TAXIIN  CRSARRTIME  ARRTIME  ARRDELAY  CRSELAPSEDTIME  ACTUALELAPSEDTIME  \
0     4.0        1655   1652.0      -3.0            90.0               87.0   
1    16.0        1426   1435.0       9.0           116.0              119.0   
2     2.0        1500   1620.0      80.0           105.0              102.0   
3     1.0        1950   1948.0      -2.0            85.0               83.0   
4     5.0        1750   1747.0      -3.0           115.0              114.0   

  CANCELLED DIVERTED   DISTANCE  
0         F    False  580 miles  
1     False        F  744 miles  
2         F    False  718 miles  
3         0        0  487 miles  
4         F    False  744 miles

Zestaw danych flights jest problematyzny z wielu względów, ale zanim na nie odpowiemy, przyjrzymy się nazwom zmiennych.

- **TRANSACTIONID**: Unikalny identyfikator

- **FLIGHTDATE**: Data lotu. Wygląda na to, że jest zakodowana jako liczba zamiast obiektu daty

- **TAILNUM**: Wygląda na to, że w niektórych wierszach zawiera @@ 

- **ORIGAIRPORTNAME** oraz **DESTAIRPORTNAME**: Wygląda na to, że nazwa miasta i stan są połączone i dołączone przed rzeczywistą nazwą lotniska 

- **CRSDEPTIME** oraz **DEPTIME**: Wygląda na to, że zawierają niepoprawnie sformatowane czasy. Wydaje się też, że: **CRSDEPTTIME** + **DEPDELAY** = **DEPTIME**

- **DEPDELAY**: Opóźnienie wyjazdu w minutach?

- **TAXIOUT**: Ile czasu upłynęło od wystartowania silnika do oderwania się od ziemi? Poza tym: **DEPTIME** + **TAXIOUT** = **WHEELSOFF**

- **WHEELSOFF**: Czas, w którym koła oderwały się od ziemi

- **WHEELSON**: Czas, w którym koła dotknęły ziemi podczas lądowania

- **TAXIIN**: Wygląda na liczbę minut od zetknięcia kół z ziemią do „parkowania”

- **CRSARRTIME**: Oczekiwany czas przylotu w formacie 24h

- **ARRTIME**: Faktyczny czas przylotu 

- **ARRDELAY**: Różnica pomiędzy **CRSARRTIME** oraz **ARRTIME**

- **CRSELAPSEDTIME**: Planowany czas podróży (minuty)

- **ACTUALELAPSEDTIME**: Rzeczywisty czas podróży (minuty)

- **CANCELLED**: Czy lot został odwołany czy nie. Niektóre wartości logiczne zostały przedstawione jako False, inne jako 0. Podobnie jest z wartościami True i 1.

- **DIVERTED**: Czy samolot został przekierowany. Podobne problemy dotyczące True/False jak powyżej?

- **DISTANCE**: Odległość (całkowita) przebyta przez samolot, zakodowana jako ciąg znaków z połączonymi z nim „milami” 

In [ ]:
flights_df

Tak więc przechodzenie przez każdą kolumnę identyfikuje problemy, które może mieć zestaw danych. Najpierw zajmijmy się kwestią **odległości**. Wygląda na to, że ma to być zakodowane _całkowite_ (zauważ, że na początku tego wykładu powiedziałem, że odległości mogą być zmiennoprzecinkowe. Dlaczego zmiana zdania?) i mają dopisane 'mile' na końcu numer. Możemy zrobić o wiele więcej rzeczy z danymi liczbowymi niż danymi tekstowymi, które reprezentują liczby, więc najpierw przekonwertujmy to na int.
Aalizująć tabelę kolumna po kolumnie dostrzegamy różne problemy, z jakimi musimy się zmierzyć w tym zbiorze danych. Można zacząć od problemu odległości. Wygląda na to, że odległość miała być zakodowana jako integery, ale pojawiają oznaczenia miar. Dane numeryczne pozwalają na o wiele więcej operacji niż dane tekstowe, dlatego należy wyczyścić te wartości tekstowe.

<table>
    <tr>
        <td> </td>
        <td><b>OrderID</b></td>
        <td><b>Cost</b></td>
        <td><b>Quantity</b></td>
        <td><b>Address</b></td>
    </tr>
    <tr>
        <td>0</td>
        <td>1234</td>
        <td>£1000.00</td>
        <td>10</td>
        <td>123 Fake Street</td>
    </tr>
    <tr>
        <td>1</td>
        <td>7890</td>
        <td>£35.50</td>
        <td>3</td>
        <td>789 Real Road</td>
    </tr>
    
</table>
W powyższej tabeli widzimy, że koszt powinien być liczbą zmiennoprzecinkową - jednak ma dołączony symbol £. Aby użyć tej kolumny jako float, musimy usunąć £. Jednak zanim to zrobimy, spójrzmy na typy danych kolumn. Odbywa się to poprzez wywołanie atrybutu `.dtypes` w naszej ramce danych. W powyższym przykładzie zwrócilibyśmy:


<table>
    <tr>
        <td>OrderID</td>
        <td>int64</td>
    </tr>
    <tr>
        <td>Cost</td>
        <td>object</td>
    </tr>
    <tr>
        <td>Quantity</td>
        <td>int64</td>
    </tr>
    <tr>
        <td>Address</td>
        <td>object</td>
    </tr>
</table>

Możemy również użyć metody `.info()`, która zwraca nam również informacje o wartościach null w każdej kolumnie (wkrótce omówimy, jak radzić sobie z wartościami null/brakującymi wartościami)

In [ ]:
## Znajdź typy obiektów dla każdej kolumny za pomocą .dtypes
flights_df.dtypes

TRANSACTIONID          int64
FLIGHTDATE             int64
AIRLINECODE           object
AIRLINENAME           object
TAILNUM               object
FLIGHTNUM              int64
ORIGINAIRPORTCODE     object
ORIGAIRPORTNAME       object
ORIGINCITYNAME        object
ORIGINSTATE           object
ORIGINSTATENAME       object
DESTAIRPORTCODE       object
DESTAIRPORTNAME       object
DESTCITYNAME          object
DESTSTATE             object
DESTSTATENAME         object
CRSDEPTIME             int64
DEPTIME              float64
DEPDELAY             float64
TAXIOUT              float64
WHEELSOFF            float64
WHEELSON             float64
TAXIIN               float64
CRSARRTIME             int64
ARRTIME              float64
ARRDELAY             float64
CRSELAPSEDTIME       float64
ACTUALELAPSEDTIME    float64
CANCELLED             object
DIVERTED              object
DISTANCE               int64
dtype: object

In [ ]:
## Znajdź typy obiektów i liczbę wartości null za pomocą .info()
flights_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1191805 entries, 0 to 1191804
Data columns (total 31 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   TRANSACTIONID      1191805 non-null  int64  
 1   FLIGHTDATE         1191805 non-null  int64  
 2   AIRLINECODE        1191805 non-null  object 
 3   AIRLINENAME        1191805 non-null  object 
 4   TAILNUM            1034988 non-null  object 
 5   FLIGHTNUM          1191805 non-null  int64  
 6   ORIGINAIRPORTCODE  1191805 non-null  object 
 7   ORIGAIRPORTNAME    1191805 non-null  object 
 8   ORIGINCITYNAME     1191805 non-null  object 
 9   ORIGINSTATE        1180963 non-null  object 
 10  ORIGINSTATENAME    1180963 non-null  object 
 11  DESTAIRPORTCODE    1191805 non-null  object 
 12  DESTAIRPORTNAME    1191805 non-null  object 
 13  DESTCITYNAME       1191805 non-null  object 
 14  DESTSTATE          1180967 non-null  object 
 15  DESTSTATENAME      1180967 non-n

If we were to sum our above <b>cost</b> column (`sales['cost'].sum()`), something akin to the following would be returned:
```£1000.00£35.50£46.10£76.35```...

Obviously this isn't what we want.. we'd rather have all our costs summed.

Try the same with the 'DISTANCE' column with our flights data

Gdybyśmy mieli zsumować powyższą kolumnę <b>cost</b> (`sales['cost'].sum()`), zwrócone zostałoby coś podobnego do następującego ciągu:
```£1000,00£35.50£46.10£76.35```...

Oczywiście nie o to nam chodzi... wolimy zsumować wszystkie nasze koszty.

Powtórz ten krok z kolumną „DISTANCE” z danymi dotyczącymi lotów

In [ ]:
## Zsumuj pierwsze 10 wystąpień kolumny „DISTANCE” w danych lotów.
# Bądź świadomy tego, gdzie używasz slicingu :). Jaka jest techniczna różnica między slicingiem przed .sum() i po?
flights_df['DISTANCE'][:10].sum()

'580 miles744 miles718 miles487 miles744 miles289 miles569 miles1240 miles223 miles677 miles'

Aby rozwiązać problem z naszymi danymi sprzedaży, musimy zrobić dwie rzeczy:
1. Usuąć „£”
2. Zamienić kolumnę na zmiennoprzecinkowy typ danych

Odbywałoby się to w następujący sposób:
```python
sales['cost'] = sales['cost'].str.strip('£')
sales['cost'] = sales['cost'].astype('float64')
```

Uzbrojeni w tę wiedzę, zamieńmy kolumnę odległości na int!

In [ ]:
## Usuń „mile” z ramki danych
flights_df["DISTANCE"] = flights_df["DISTANCE"].str.strip(" miles")
## Zamień kolumnę na typ int64
flights_df["DISTANCE"] = flights_df["DISTANCE"].astype("int64")
## Sprawdź, czy kolumna została pomyślnie przekonwertowana na int

Świetnie! W ten sposób możemy przekonwertować nieuporządkowane dane tekstowe na liczby. Przyjrzyjmy się teraz konwersji danych na wartości kategoryczne.

W naszym zbiorze danych mamy wiele kolumn, które mogą być kategoryczne. Czy potrafisz określić, które to są?
<br>
<details>
    <summary><b>></b> Categorical variables (click to reveal)</summary>
    <ul>
        <li>AIRLINECODE</li>
        <li>AIRLINENAME</li>
        <li>ORIGINAIRPORTCODE</li>
        <li>ORIGAIRPORTNAME</li>
        <li>ORIGINCITYNAME</li>
        <li>ORIGINSTATE</li>
        <li>ORIGINSTATENAME</li>
        <li>DESTAIRPORTCODE</li>
        <li>DESTAIRPORTNAME</li>
        <li>DESTCITYNAME</li>
        <li>DESTSTATE</li>
        <li>DESTSTATENAME</li>
    </ul>
</details>

Używając metody `.describe()` możemy dostać więcej informacji o konkretnej kolumnie. Użyjmy jako przykładu `AIRLINECODE`

In [ ]:
flights_df['DISTANCE'].describe()

count    1.191805e+06
mean     7.336410e+02
std      5.647814e+02
min      1.100000e+01
25%      3.250000e+02
50%      5.810000e+02
75%      9.660000e+02
max      4.983000e+03
Name: DISTANCE, dtype: float64

In [ ]:
flights_df['AIRLINECODE']

Gdy uruchomimy `.describe()` nad tą zmienną, otrzymujemy kilka (średnio) użytecznych statystyk. Widzimy jednak, że typ danych tej kolumny został zinterpretowany jako `object`. Z tabeli typów danych, którą wprowadziliśmy wcześniej, widzimy, że istnieje obsługa kategorii. Przekształćmy to w kategorię i zobaczmy różnicę w stosunku do metody opisu.

In [ ]:
flights_df['AIRLINECODE'] = flights_df['AIRLINECODE'].astype('category')
flights_df['AIRLINECODE']

0          WN
1          CO
2          WN
3          WN
4          CO
           ..
1191800    EV
1191801    DL
1191802    DL
1191803    EV
1191804    EV
Name: AIRLINECODE, Length: 1191805, dtype: category
Categories (26, object): ['9E', 'AA', 'AS', 'B6', ..., 'VX', 'WN', 'XE', 'YV']

In [ ]:
flights_df['AIRLINECODE'].describe()

Tak naprawdę brak widocznej różnicy po użyciu metody `.describe()` (co dziwne, wciąż zwraca dtype object)! Użyjemy metody `.info()`, aby sprawdzić nasze zużycie pamięci

In [ ]:
flights_df['AIRLINECODE'] = flights_df['AIRLINECODE'].astype('object')
flights_df.info()

In [ ]:
flights_df['AIRLINECODE'] = flights_df['AIRLINECODE'].astype('category')
flights_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1191805 entries, 0 to 1191804
Data columns (total 32 columns):
 #   Column             Non-Null Count    Dtype   
---  ------             --------------    -----   
 0   TRANSACTIONID      1191805 non-null  int64   
 1   FLIGHTDATE         1191805 non-null  int64   
 2   AIRLINECODE        1191805 non-null  category
 3   AIRLINENAME        1191805 non-null  object  
 4   TAILNUM            1034988 non-null  object  
 5   FLIGHTNUM          1191805 non-null  int64   
 6   ORIGINAIRPORTCODE  1191805 non-null  object  
 7   ORIGAIRPORTNAME    1191805 non-null  object  
 8   ORIGINCITYNAME     1191805 non-null  object  
 9   ORIGINSTATE        1180963 non-null  object  
 10  ORIGINSTATENAME    1180963 non-null  object  
 11  DESTAIRPORTCODE    1191805 non-null  object  
 12  DESTAIRPORTNAME    1191805 non-null  object  
 13  DESTCITYNAME       1191805 non-null  object  
 14  DESTSTATE          1180967 non-null  object  
 15  DESTSTATENAME  

Widzimy, że nasze zużycie pamięci spadło o około 8 MB po przekształceniu tej jednej kolumny w kategorię! Ok, tak, wprawdzie nie jest to wielka sprawa przy pracy z danymi o tym rozmiarze, ale pamiętaj, że oszczędność pamięci pochodzi tylko z jednej z wielu kategorycznych kolumn, które mamy.

Więc dlaczego tak jest? Cóż, pod maską Pandas reprezentują kategorie jako typy całkowite. W rzeczywistości coś, na co możesz natknąć się podczas pracy z innymi zestawami danych, to jawne wyświetlanie kolumny kategorii zakodowanej jako liczby całkowite. Zmodyfikujmy naszą ramkę danych, aby zobaczyć, co się stanie dalej.

In [ ]:
flights_df['AIRLINECODE_ASINT'] = flights_df['AIRLINECODE'].cat.codes.astype('int8')
flights_df['AIRLINECODE_ASINT']

0          23
1           4
2          23
3          23
4           4
           ..
1191800     7
1191801     6
1191802     6
1191803     7
1191804     7
Name: AIRLINECODE_ASINT, Length: 1191805, dtype: int8

Kiedy uruchamiamy `.describe()` możemy zobaczyć zwrócone statystyki, które nie mają sensu dla naszej kolumny:

In [ ]:
flights_df['AIRLINECODE_ASINT'].describe()

count    1.191805e+06
mean     1.305387e+01
std      8.211102e+00
min      0.000000e+00
25%      6.000000e+00
50%      1.500000e+01
75%      2.100000e+01
max      2.500000e+01
Name: AIRLINECODE_ASINT, dtype: float64

Nie ma sensu, aby kolumna kategoryczna miała średnią lub jakiekolwiek inne właściwości statystyczne.

In [ ]:
flights_df = flights_df.drop('AIRLINECODE_ASINT', 1)

Jakie typy danych (numeryczne, datetime, tekst lub kategoryczne) przypisałbyś następującym danym?:

- Opis przedmiotu
- Roczny dochód
- Rozmiar odzieży
- Czas przylotu samolotu
- Urodziny
- Smaki koktajli mlecznych w McDonalds
- Pierwsza połowa kodu pocztowego
- Pełny kod pocztowy
- Czas potrzebny biegaczom na ukończenie 5K

## Zduplikowane wartości

Innym częstym problemem, z którym możemy się spotkać, są **zduplikowane wartości**. Jak sama nazwa wskazuje, dzieje się tak, gdy te same wartości powtarzają się w wielu wierszach lub kolumnach:
<table>
    <tr>
        <td><b>first_name</b></td>
        <td><b>last_name</b></td>
        <td><b>address</b></td>
        <td><b>age</b></td>
        <td><b>income</b></td>
    </tr>
    <tr>
        <td>John</td>
        <td>Doe</td>
        <td>123 Real Street</td>
        <td>25</td>
        <td>£28000</td>
    </tr>
    <tr>
        <td>Jane</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td>29</td>
        <td>£32000</td>
    </tr>
    <tr>
        <td>Jane</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td>29</td>
        <td>£32000</td>
    </tr>
    <tr>
        <td>Mark</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td>31</td>
        <td>£32000</td>
    </tr>
</table>

W powyższym przykładzie widzimy, że Jane Smith ma dwa wpisy bezpośrednio zduplikowane. Jednak w niektórych przypadkach możemy zobaczyć bardzo podobne wpisy:

<table>
    <tr>
        <td><b>first_name</b></td>
        <td><b>last_name</b></td>
        <td><b>address</b></td>
        <td><b>age</b></td>
        <td><b>income</b></td>
    </tr>
    <tr>
        <td>John</td>
        <td>Doe</td>
        <td>123 Real Street</td>
        <td>25</td>
        <td>£28000</td>
    </tr>
    <tr>
        <td>Jane</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td><b>28</b></td>
        <td>£32000</td>
    </tr>
    <tr>
        <td>Jane</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td>29</td>
        <td>£32000</td>
    </tr>
    <tr>
        <td>Mark</td>
        <td>Smith</td>
        <td>789 Fake Road</td>
        <td>31</td>
        <td>£32000</td>
    </tr>
</table>

(Różnica wieku między obiema Jane Smith). Ten rodzaj duplikatu błędu jest najprawdopodobniej spowodowany problemem z wprowadzaniem danych lub ponownym przesłaniem dowolnego formularza złożonego przez Jane – który został wprowadzony do bazy danych bez usuwania jej starego wpisu.

Najczęściej jednak duplikaty danych wynikają z błędów/wzorców projektowych w potokach danych lub najczęściej z łączenia baz danych i konsolidacji danych z różnych zestawów danych/baz danych, które mogą zachować zduplikowane wartości.

Pandas udostępnia nam metodę [`.duplicated()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html). Użyjmy tego w naszej ramce danych, aby zobaczyć, co zwraca

In [ ]:
flights_df.duplicated()

0          False
1          False
2          False
3          False
4          False
           ...  
1191800    False
1191801    False
1191802    False
1191803    False
1191804    False
Length: 1191805, dtype: bool

Zauważ, że możemy użyć `.sum()` do wartości logicznych. Zasadniczo False są interpretowane jako 0, a True jako 1. Tak więc sumując ramkę danych, możemy uzyskać całkowitą liczbę zduplikowanych wartości!

In [ ]:
flights_df.duplicated().sum()

0

Brak duplikatów to dobry wynik, ale nie dajmy się zwieść. Przypomnijmy sobie duplikat Jane Smith powyżej. Zduplikowana metoda nie zwróciłaby wartości true, ponieważ cały wiersz nie był dokładnym duplikatem. Dlatego możemy w metodzie `.duplicated()` skorzystać z dwóch argumentów: `subset` i `keep`. Dla argumentu subset podaj listę nazw kolumn, w których chcemy sprawdzić duplikaty, a w argumencie keep podaj  1 z 3 wartości: „pierwsza”, „ostatnia” lub „Fałsz”. Z dokumentacji wiemy, że:
- `first` : Oznacz duplikaty jako True z wyjątkiem pierwszego wystąpienia.
- `last` : Zaznacz duplikaty jako True z wyjątkiem ostatniego wystąpienia.
- `False` : Oznacz wszystkie duplikaty jako True.

W wielu przypadkach wybranie podzbioru jest bardziej intuicyjne niż naukowe.

In [ ]:
## Znajdź duplikaty w ramce danych lotów w następujących kolumnach, ustaw keep = False:
 # "ORIGAIRPORTNAME", "DESTAIRPORTNAME", "AIRLINECODE", "FLIGHTDATE", "CRSDEPTIME", "DEPTIME", "ARRTIME"
 # Przypisz to zmiennej 'duplicates'
 # Czy wybrałam dobre klolumny? Postąpiłbyś inaczej?
duplicates = flights_df.duplicated(subset=["ORIGAIRPORTNAME", "DESTAIRPORTNAME", "AIRLINECODE", "FLIGHTDATE", "CRSDEPTIME", "DEPTIME", "ARRTIME"], keep=False)
    
# Używając df[duplikaty], zwracane są punkty danych, w których istnieją duplikaty.
## Zwróć duplikaty dla ramki danych lotów
flights_df[duplicates]

TRANSACTIONID  FLIGHTDATE AIRLINECODE                 AIRLINENAME  \
34300          1974100    19920528          AS    Alaska Airlines Inc.: AS   
103931        32273300    19980108          AA  American Airlines Inc.: AA   
125014        22096600    19960118          AA  American Airlines Inc.: AA   
125015        22097200    19960118          AA  American Airlines Inc.: AA   
162444        27454200    19970113          AA  American Airlines Inc.: AA   
162445        27453500    19970113          AA  American Airlines Inc.: AA   
192865        21755500    19960108          AA  American Airlines Inc.: AA   
377880        21910100    19960107          AA  American Airlines Inc.: AA   
377882        21910200    19960107          AA  American Airlines Inc.: AA   
453974         1977500    19920528          AS    Alaska Airlines Inc.: AS   
583756        27175600    19970126          AA  American Airlines Inc.: AA   
585484        21808100    19960111          AA  American Airlines Inc.: AA   
585779        20837400    19960108          AA  American Airlines Inc.: AA   
603812        21095900    19960108          AA  American Airlines Inc.: AA   
603813        21097100    19960108          AA  American Airlines Inc.: AA   
760553        32274400    19980108          AA  American Airlines Inc.: AA   
769566        22395100    19960107          AA  American Airlines Inc.: AA   
769567        22394500    19960107          AA  American Airlines Inc.: AA   
810576        21756100    19960108          AA  American Airlines Inc.: AA   
1174655       31977900    19980106          AA  American Airlines Inc.: AA   
1174656       31977700    19980106          AA  American Airlines Inc.: AA   
1175821       20838000    19960108          AA  American Airlines Inc.: AA   
1175879       21808800    19960111          AA  American Airlines Inc.: AA   
1185114       27174500    19970126          AA  American Airlines Inc.: AA   

        TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
34300       NaN         88               ANC   
103931   UNKNOW        496               LAX   
125014   UNKNOW       1388               MCI   
125015   UNKNOW       1988               MCI   
162444   UNKNOW       2087               IAH   
162445   UNKNOW       2056               IAH   
192865   UNKNOW        613               LGA   
377880   UNKNOW        883               DCA   
377882   UNKNOW       1009               DCA   
453974      NaN        118               ANC   
583756   UNKNOW       2085               ORD   
585484   UNKNOW       1311               ORD   
585779   UNKNOW        218               ORD   
603812   UNKNOW        236               DFW   
603813   UNKNOW       1684               DFW   
760553   UNKNOW       2098               LAX   
769566   UNKNOW         32               LAX   
769567   UNKNOW          2               LAX   
810576   UNKNOW       1485               LGA   
1174655  UNKNOW        386               ORD   
1174656  UNKNOW        378               ORD   
1175821  UNKNOW       1920               ORD   
1175879  UNKNOW       1995               ORD   
1185114  UNKNOW       1140               ORD   

                                           ORIGAIRPORTNAME     ORIGINCITYNAME  \
34300     AnchorageAK: Ted Stevens Anchorage International          Anchorage   
103931            Los AngelesCA: Los Angeles International        Los Angeles   
125014            Kansas CityMO: Kansas City International        Kansas City   
125015            Kansas CityMO: Kansas City International        Kansas City   
162444     HoustonTX: George Bush Intercontinental/Houston            Houston   
162445     HoustonTX: George Bush Intercontinental/Houston            Houston   
192865                               New YorkNY: LaGuardia           New York   
377880     WashingtonDC: Ronald Reagan Washington National         Washington   
377882     WashingtonDC: Ronald Reagan Washington National         Washington   
453974    AnchorageAK: Ted Stevens Anchorage In

Jako drugorzędną obserwację widzimy, że „TALINUM” również przyjmuje wartość „UNKNOWN” dla brakujących wartości. Zanotujemy to, abyśmy mogli zająć się tym później.

Aby posortować naszą ramkę danych, możemy użyć metody [`.sort_values()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html). Przeczytaj dokumentację i użyj tej metody, aby posortować ramkę danych według nazwy kolumny, która Twoim zdaniem jest odpowiednia (taka, która pozwala łatwo zweryfikować, czy zwrócone wpisy są faktycznymi duplikatami).

In [ ]:
## Posortuj zduplikowane wartości według odpowiedniego indeksu
flights_df[duplicates].sort_values("FLIGHTDATE")


TRANSACTIONID  FLIGHTDATE AIRLINECODE                 AIRLINENAME  \
34300          1974100    19920528          AS    Alaska Airlines Inc.: AS   
453974         1977500    19920528          AS    Alaska Airlines Inc.: AS   
769567        22394500    19960107          AA  American Airlines Inc.: AA   
769566        22395100    19960107          AA  American Airlines Inc.: AA   
377880        21910100    19960107          AA  American Airlines Inc.: AA   
377882        21910200    19960107          AA  American Airlines Inc.: AA   
1175821       20838000    19960108          AA  American Airlines Inc.: AA   
810576        21756100    19960108          AA  American Airlines Inc.: AA   
603813        21097100    19960108          AA  American Airlines Inc.: AA   
192865        21755500    19960108          AA  American Airlines Inc.: AA   
603812        21095900    19960108          AA  American Airlines Inc.: AA   
585779        20837400    19960108          AA  American Airlines Inc.: AA   
585484        21808100    19960111          AA  American Airlines Inc.: AA   
1175879       21808800    19960111          AA  American Airlines Inc.: AA   
125015        22097200    19960118          AA  American Airlines Inc.: AA   
125014        22096600    19960118          AA  American Airlines Inc.: AA   
162445        27453500    19970113          AA  American Airlines Inc.: AA   
162444        27454200    19970113          AA  American Airlines Inc.: AA   
583756        27175600    19970126          AA  American Airlines Inc.: AA   
1185114       27174500    19970126          AA  American Airlines Inc.: AA   
1174655       31977900    19980106          AA  American Airlines Inc.: AA   
1174656       31977700    19980106          AA  American Airlines Inc.: AA   
760553        32274400    19980108          AA  American Airlines Inc.: AA   
103931        32273300    19980108          AA  American Airlines Inc.: AA   

        TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
34300       NaN         88               ANC   
453974      NaN        118               ANC   
769567   UNKNOW          2               LAX   
769566   UNKNOW         32               LAX   
377880   UNKNOW        883               DCA   
377882   UNKNOW       1009               DCA   
1175821  UNKNOW       1920               ORD   
810576   UNKNOW       1485               LGA   
603813   UNKNOW       1684               DFW   
192865   UNKNOW        613               LGA   
603812   UNKNOW        236               DFW   
585779   UNKNOW        218               ORD   
585484   UNKNOW       1311               ORD   
1175879  UNKNOW       1995               ORD   
125015   UNKNOW       1988               MCI   
125014   UNKNOW       1388               MCI   
162445   UNKNOW       2056               IAH   
162444   UNKNOW       2087               IAH   
583756   UNKNOW       2085               ORD   
1185114  UNKNOW       1140               ORD   
1174655  UNKNOW        386               ORD   
1174656  UNKNOW        378               ORD   
760553   UNKNOW       2098               LAX   
103931   UNKNOW        496               LAX   

                                           ORIGAIRPORTNAME     ORIGINCITYNAME  \
34300     AnchorageAK: Ted Stevens Anchorage International          Anchorage   
453974    AnchorageAK: Ted Stevens Anchorage International          Anchorage   
769567            Los AngelesCA: Los Angeles International        Los Angeles   
769566            Los AngelesCA: Los Angeles International        Los Angeles   
377880     WashingtonDC: Ronald Reagan Washington National         Washington   
377882     WashingtonDC: Ronald Reagan Washington National         Washington   
1175821            ChicagoIL: Chicago O'Hare International            Chicago   
810576                               New YorkNY: LaGuardia           New York   
603813   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...  Dallas/Fort Worth   
192865                               New YorkNY

In [ ]:
flights_df["TAILNUM"].isna().sum()

156817

### Radzenie sobie z duplikatami

Mamy dwie opcje:
1. Uśrednienie wartości, tam gdzie to możliwe
2. Usunięcie jednego z duplikatów (lub wielu, tak by pozostał pojedynczy wiersz)


##### Uśrednianie

Uśrednianie po zduplikowanych wartościach można tak naprawdę wykonać tylko na typach danych, które mają sens. W powyższej tabeli pierwsze dwa wpisy mają prawidłowe czasy, które możemy uśrednić. Ogólnie rzecz biorąc, sposób, w jaki uśredniamy, to grupowanie według odpowiednich kolumn (poprzez `.groupby()`) i łączenie tego z funkcją `.agg()`. W tym przypadku chcemy pogrupować wg kolumn w podzbiorze poza kolumnami, które nas interesują (np. wg. czasu). Naszym argumentem do `.agg()` jest słownik z parami klucz-wartość nazw kolumn i funkcją agregacji, którą chcemy nad nimi zastosować (np. suma, różnica, średnia itp.).

In [ ]:
summaries = {"CRSARRTIME": "mean", "ARRTIME": "mean", "ARRDELAY": "mean", "CRSELAPSEDTIME": "mean", "ACTUALELAPSEDTIME": "mean"}

grouped_duplicates = flights_df[duplicates].groupby(["FLIGHTDATE", "AIRLINECODE", "ORIGAIRPORTNAME", "DESTAIRPORTNAME"])
grouped_duplicates_min_transactionid = grouped_duplicates["TRANSACTIONID"].min().reset_index()

f_df_duplicates = pd.merge(
    grouped_duplicates_min_transactionid,
    grouped_duplicates.agg(summaries).reset_index(),
    how="inner"
).sort_values("TRANSACTIONID")

f_df_duplicates


FLIGHTDATE AIRLINECODE  \
151      19920528          AS   
3827     19960108          AA   
3842     19960108          AA   
3873     19960108          AA   
5701     19960111          AA   
...           ...         ...   
16843    19980108          YV   
16844    19980108          YV   
16845    19980108          YV   
16846    19980108          YV   
16847    19980108          YV   

                                         ORIGAIRPORTNAME  \
151     AnchorageAK: Ted Stevens Anchorage International   
3827             ChicagoIL: Chicago O'Hare International   
3842   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...   
3873                               New YorkNY: LaGuardia   
5701             ChicagoIL: Chicago O'Hare International   
...                                                  ...   
16843    WashingtonDC: Ronald Reagan Washington National   
16844    WashingtonDC: Ronald Reagan Washington National   
16845    WashingtonDC: Ronald Reagan Washington National   
16846    WashingtonDC: Ronald Reagan Washington National   
16847    WashingtonDC: Ronald Reagan Washington National   

                                        DESTAIRPORTNAME  TRANSACTIONID  \
151             SeattleWA: Seattle/Tacoma International      1974100.0   
3827                  HartfordCT: Bradley International     20837400.0   
3842    WashingtonDC: Ronald Reagan Washington National     21095900.0   
3873                       MiamiFL: Miami International     21755500.0   
5701   MinneapolisMN: Minneapolis-St Paul International     21808100.0   
...                                                 ...            ...   
16843  MinneapolisMN: Minneapolis-St Paul International            NaN   
16844         New YorkNY: John F. Kennedy International            NaN   
16845                             New YorkNY: LaGuardia            NaN   
16846           SeattleWA: Seattle/Tacoma International            NaN   
16847   WashingtonDC: Ronald Reagan Washington National            NaN   

       CRSARRTIME  ARRTIME  ARRDELAY  CRSELAPSEDTIME  ACTUALELAPSEDTIME  
151        1405.0   1405.0       0.0           195.0              195.0  
3827          0.0      NaN       NaN             NaN                NaN  
3842          0.0      NaN       NaN             NaN                NaN  
3873          0.0      NaN       NaN             NaN                NaN  
5701          0.0      NaN       NaN             NaN                NaN  
...           ...      ...       ...             ...                ...  
16843         NaN      NaN       NaN             NaN                NaN  
16844         NaN      NaN       NaN             NaN                NaN  
16845         NaN      NaN       NaN             NaN                NaN  
16846         NaN      NaN       NaN             NaN                NaN  
16847         NaN      NaN       NaN             NaN                NaN  

[16848 rows x 10 columns]

In [ ]:
# Dlaczego w polu TRANSACTIONID jest teraz tak wiele nowych NaN?
## Jak się ich pozbyć?
f_df_duplicates = f_df_duplicates.dropna(subset=["TRANSACTIONID"])

## Ponownie zakoduj TRANSACTIONID do int64
f_df_duplicates["TRANSACTIONID"] = f_df_duplicates["TRANSACTIONID"].astype("int64")
f_df_duplicates

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


FLIGHTDATE AIRLINECODE  \
151      19920528          AS   
3827     19960108          AA   
3842     19960108          AA   
3873     19960108          AA   
5701     19960111          AA   
2007     19960107          AA   
7597     19960118          AA   
1994     19960107          AA   
11317    19970126          AA   
9460     19970113          AA   
13191    19980106          AA   
15093    19980108          AA   

                                         ORIGAIRPORTNAME  \
151     AnchorageAK: Ted Stevens Anchorage International   
3827             ChicagoIL: Chicago O'Hare International   
3842   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...   
3873                               New YorkNY: LaGuardia   
5701             ChicagoIL: Chicago O'Hare International   
2007     WashingtonDC: Ronald Reagan Washington National   
7597            Kansas CityMO: Kansas City International   
1994            Los AngelesCA: Los Angeles International   
11317            ChicagoIL: Chicago O'Hare International   
9460     HoustonTX: George Bush Intercontinental/Houston   
13191            ChicagoIL: Chicago O'Hare International   
15093           Los AngelesCA: Los Angeles International   

                                         DESTAIRPORTNAME  TRANSACTIONID  \
151              SeattleWA: Seattle/Tacoma International        1974100   
3827                   HartfordCT: Bradley International       20837400   
3842     WashingtonDC: Ronald Reagan Washington National       21095900   
3873                        MiamiFL: Miami International       21755500   
5701    MinneapolisMN: Minneapolis-St Paul International       21808100   
2007             ChicagoIL: Chicago O'Hare International       21910100   
7597   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...       22096600   
1994           New YorkNY: John F. Kennedy International       22394500   
11317   MinneapolisMN: Minneapolis-St Paul International       27174500   
9460   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...       27453500   
13191                              New YorkNY: LaGuardia       31977700   
15093            ChicagoIL: Chicago O'Hare International       32273300   

       CRSARRTIME  ARRTIME  ARRDELAY  CRSELAPSEDTIME  ACTUALELAPSEDTIME  
151        1405.0   1405.0       0.0           195.0              195.0  
3827          0.0      NaN       NaN             NaN                NaN  
3842          0.0      NaN       NaN             NaN                NaN  
3873          0.0      NaN       NaN             NaN                NaN  
5701          0.0      NaN       NaN             NaN                NaN  
2007          0.0      NaN       NaN             NaN                NaN  
7597          0.0      NaN       NaN             NaN                NaN  
1994          0.0      NaN       NaN             NaN                NaN  
11317         0.0      NaN       NaN             NaN                NaN  
9460          0.0      NaN       NaN             NaN                NaN  
13191         0.0      NaN       NaN             NaN                NaN  
15093         0.0      NaN       NaN             NaN                NaN

In [ ]:
# Metoda .update() pozwala nam aktualizować rekordy w jednej ramce danych na podstawie wartości w innej
# Potrzebny jest pewien sposób "powiązania" rekordów do nadpisania/aktualizacji, jeśli nie chcemy używać domyślnego indeksu ramki danych
## Tak więc, używając metody .set_index(), ustaw ramkom flight_df i f_df_duplicates nowy indeks na unikalny klucz indentifera, który obaj współdzielą
flights_df = flights_df.set_index("TRANSACTIONID")
f_df_duplicates = f_df_duplicates.set_index("TRANSACTIONID")

# Teraz możemy zaktualizować ramkę flight_df o nową ramkę danych
flights_df.update(f_df_duplicates)

## I na koniec możemy opcjonalnie zresetować indeks, aby uzyskać domyślne indeksowanie ramki danych
flights_df=flights_df.reset_index()

In [ ]:
flights_df.head(2)

TRANSACTIONID  FLIGHTDATE AIRLINECODE                     AIRLINENAME  \
0       54548800  20020101.0          WN      Southwest Airlines Co.: WN   
1       55872300  20020101.0          CO  Continental Air Lines Inc.: CO   

  TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
0  N103@@       1425               ABQ   
1  N83872        150               ABQ   

                                    ORIGAIRPORTNAME ORIGINCITYNAME  \
0  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   
1  AlbuquerqueNM: Albuquerque International Sunport    Albuquerque   

  ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
0          NM      New Mexico             DAL   
1          NM      New Mexico             IAH   

                                   DESTAIRPORTNAME DESTCITYNAME DESTSTATE  \
0                      DallasTX: Dallas Love Field       Dallas        TX   
1  HoustonTX: George Bush Intercontinental/Houston      Houston        TX   

  DESTSTATENAME  CRSDEPTIME  DEPTIME  DEPDELAY  TAXIOUT  WHEELSOFF  WHEELSON  \
0         Texas        1425   1425.0       0.0      8.0     1433.0    1648.0   
1         Texas        1130   1136.0       6.0     12.0     1148.0    1419.0   

   TAXIIN  CRSARRTIME  ARRTIME  ARRDELAY  CRSELAPSEDTIME  ACTUALELAPSEDTIME  \
0     4.0      1655.0   1652.0      -3.0            90.0               87.0   
1    16.0      1426.0   1435.0       9.0           116.0              119.0   

  CANCELLED DIVERTED  DISTANCE  AIRLINECODE_ASINT  
0         F    False       580                 23  
1     False        F       744                  4

In [ ]:
flights_df[flights_df["TRANSACTIONID"]==1974100]

TRANSACTIONID  FLIGHTDATE AIRLINECODE               AIRLINENAME  \
34300        1974100  19920528.0          AS  Alaska Airlines Inc.: AS   

      TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
34300     NaN         88               ANC   

                                        ORIGAIRPORTNAME ORIGINCITYNAME  \
34300  AnchorageAK: Ted Stevens Anchorage International      Anchorage   

      ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
34300          AK          Alaska             SEA   

                               DESTAIRPORTNAME DESTCITYNAME DESTSTATE  \
34300  SeattleWA: Seattle/Tacoma International      Seattle        WA   

      DESTSTATENAME  CRSDEPTIME  DEPTIME  DEPDELAY  TAXIOUT  WHEELSOFF  \
34300    Washington         950    950.0       0.0      NaN        NaN   

       WHEELSON  TAXIIN  CRSARRTIME  ARRTIME  ARRDELAY  CRSELAPSEDTIME  \
34300       NaN     NaN      1405.0   1405.0       0.0           195.0   

       ACTUALELAPSEDTIME CANCELLED DIVERTED  DISTANCE  AIRLINECODE_ASINT  
34300              195.0     False        F      1448                  2

##### Usuwanie duplikatów

Jeśli chodzi o usuwanie duplikatów, Pandas udostępnia nam metodę `.drop_duplicates()`, która przyjmuje trzy argumenty:
1. `subset`
2. `keep`
3. `inplace` - wartość boolean, czy chcemy nadpisać ramkę czy nie

In [ ]:
subset = ["ORIGAIRPORTNAME", "DESTAIRPORTNAME", "AIRLINECODE", "FLIGHTDATE", "CRSDEPTIME", "DEPTIME", "ARRTIME"]
## Używając inplace = True, usuń duplikaty. Zastanów się, jaką wartość powinniśmy ustawić argumentowi keep
flights_df.drop_duplicates(subset, 'first', inplace=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop_duplicates except for the argument 'subset' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
flights_df[duplicates].tail()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


TRANSACTIONID  FLIGHTDATE AIRLINECODE                 AIRLINENAME  \
585484        21808100  19960111.0          AA  American Airlines Inc.: AA   
585779        20837400  19960108.0          AA  American Airlines Inc.: AA   
603812        21095900  19960108.0          AA  American Airlines Inc.: AA   
769566        22395100  19960107.0          AA  American Airlines Inc.: AA   
1174655       31977900  19980106.0          AA  American Airlines Inc.: AA   

        TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
585484   UNKNOW       1311               ORD   
585779   UNKNOW        218               ORD   
603812   UNKNOW        236               DFW   
769566   UNKNOW         32               LAX   
1174655  UNKNOW        386               ORD   

                                           ORIGAIRPORTNAME     ORIGINCITYNAME  \
585484             ChicagoIL: Chicago O'Hare International            Chicago   
585779             ChicagoIL: Chicago O'Hare International            Chicago   
603812   Dallas/Fort WorthTX: Dallas/Fort Worth Interna...  Dallas/Fort Worth   
769566            Los AngelesCA: Los Angeles International        Los Angeles   
1174655            ChicagoIL: Chicago O'Hare International            Chicago   

        ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
585484           IL        Illinois             MSP   
585779           IL        Illinois             BDL   
603812           TX           Texas             DCA   
769566           CA      California             JFK   
1174655          IL        Illinois             LGA   

                                          DESTAIRPORTNAME DESTCITYNAME  \
585484   MinneapolisMN: Minneapolis-St Paul International  Minneapolis   
585779                  HartfordCT: Bradley International     Hartford   
603812    WashingtonDC: Ronald Reagan Washington National   Washington   
769566          New YorkNY: John F. Kennedy International     New York   
1174655                             New YorkNY: LaGuardia     New York   

        DESTSTATE DESTSTATENAME  CRSDEPTIME  DEPTIME  DEPDELAY  TAXIOUT  \
585484         MN     Minnesota           0      NaN       NaN      NaN   
585779         CT   Connecticut           0      NaN       NaN      NaN   
603812         VA      Virginia           0      NaN       NaN      NaN   
769566         NY      New York           0      NaN       NaN      NaN   
1174655        NY      New York           0      NaN       NaN      NaN   

         WHEELSOFF  WHEELSON  TAXIIN  CRSARRTIME  ARRTIME  ARRDELAY  \
585484         NaN       NaN     NaN         0.0      NaN       NaN   
585779         NaN       NaN     NaN         0.0      NaN       NaN   
603812         NaN       NaN     NaN         0.0      NaN       NaN   
769566         NaN       NaN     NaN         0.0      NaN       NaN   
1174655        NaN       NaN     NaN         0.0      NaN       NaN   

         CRSELAPSEDTIME  ACTUALELAPSEDTIME CANCELLED DIVERTED  DISTANCE  \
585484              NaN                NaN      True    False       334   
585779              NaN                NaN      True    False       783   
603812              NaN                NaN      True    False      1192   
769566              NaN                NaN      True    False      2475   
1174655             NaN                NaN      True    False       733   

         AIRLINECODE_ASINT  
585484                   1  
585779                   1  
603812                   1  
769566                   1  
1174655                  1

## Dane kategoryczne

Wspomnieliśmy już o danych kategorycznych wcześniej, ale tutaj bardziej szczegółowo zdefiniujemy to pojęcie. Zmienne danych kategorycznych przyjmują swoją wartość z predefiniowanego zestawu kategorii. Widzieliśmy powyższy przykład z kodami AIRLINE.

Czy poniższe zmienne są kategoryczne?
- TAILNUM
- FLIGHTNUM
- ORIGINAIRPORTCODE
- ORIGAIRPORTNAME
- CANCELLED

A kolumny w poniższej tabeli?

<table>
    <tr>
        <td><b>First Name</b></td>
        <td><b>Last Name</b></td>
        <td><b>Age</b></td>
        <td><b>Address</b></td>
        <td><b>District Postcode</b></td>
        <td><b>Full Postcode</b></td>
        <td><b>Married</b><td>
    </tr>
    <tr>
        <td>John</td>
        <td>Doe</td>
        <td>31</td>
        <td>123 Fake Street, Realtown</td>
        <td>RT1</td>
        <td>RT1 3NV</td>
        <td>True</td>
    </tr>
    <tr>
        <td>Diane</td>
        <td>Smith</td>
        <td>31</td>
        <td>42 World Road, Realtown</td>
        <td>RT2</td>
        <td>RT2 7XU</td>
        <td>False</td>
    </tr>
    <tr>
        <td>Kate</td>
        <td>Doe</td>
        <td>29</td>
        <td>123 Fake Street, Realtown</td>
        <td>RT1</td>
        <td>RT1 3NV</td>
        <td>False</td>
    </tr>
    <tr>
        <td>Charlie</td>
        <td>Doe</td>
        <td>33</td>
        <td>789 Real Road, Fakecity</td>
        <td>FC2</td>
        <td>FC2 9ER</td>
        <td>True</td>        
    </tr>    
</table>
Dane kategoryczne mogą przyjmować tylko jedną ze skończonego zestawu wartości i nie jest możliwe, aby wykroczyły poza te uprzednio zdefiniowane kategorie. Jednak podczas procesu zbierania danych może wystąpić szum w naszych danych (np. jeśli nasze dane kategoryczne zostały zebrane za pomocą okienka tekstowego do wprowadzania dowolnej treści).


Istnieje kilka sposobów radzenia sobie z niespójnymi kategoriami:
1. Usuwanie danych
2. Zmiana mapowania kategorii
3. Wnioskowanie kategorii

### Usuwanie danych


Przyjrzyjmy się kolumnie `ORIGINSTATENAME`. Usunięcie danych jest wymagane, gdy mamy wartość, która w naszym wpisie nie znajduje się (koncepcyjnie) we wstępnie zdefiniowanym zestawie kategorii. Zaczniemy od zwrócenia wszystkich unikalnych wartości w zmiennej.

In [ ]:
## skonstruuj zbiór unikalnych wartości w ORIGINSTATENAME
states = set(flights_df["ORIGINSTATENAME"])
states

{'Alabama',
 'Alaska',
 'Arizona',
 'Arkansas',
 'California',
 'Colorado',
 'Connecticut',
 'Delaware',
 'Florida',
 'Georgia',
 'Hawaii',
 'Idaho',
 'Illinois',
 'Indiana',
 'Iowa',
 'Kentucky',
 'Louisiana',
 'Maine',
 'Maryland',
 'Massachusetts',
 'Michigan',
 'Minnesota',
 'Mississippi',
 'Missouri',
 'Montana',
 'Nebraska',
 'Nevada',
 'New Hampshire',
 'New Jersey',
 'New Mexico',
 'New York',
 'North Carolina',
 'North Dakota',
 'Ohio',
 'Oregon',
 'Pennsylvania',
 'Puerto Rico',
 'Rhode Island',
 'South Carolina',
 'South Dakota',
 'Tennessee',
 'Texas',
 'U.S. Pacific Trust Territories and Possessions',
 'U.S. Virgin Islands',
 'Utah',
 'Vermont',
 'Virginia',
 'Washington',
 'West Virginia',
 'Wisconsin',
 'Wyoming',
 nan}

Załóżmy teraz, że otrzymaliśmy kilka nowych wpisów, których nazwy stanów nie występują w tym predefiniowanym zestawie kategorii (na przykład „Fakestate”).

In [ ]:
## używając metody .at() lub .iat() na flight_df, zmodyfikuj jeden z wierszy
## w naszej tabeli, aby mieć ORIGINSTATENAME jako Fakestate
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html
flights_df.at[len(flights_df), "ORIGINSTATENAME"] = "Fakestate"

In [ ]:
# Wyświetl unikalne wpisy w ORIGINSTATENAME w naszej zmodyfikowanej ramce danych
flights_df["ORIGINSTATENAME"].unique()
# LUB set(flights_df["ORIGINSTATENAME"])

array(['New Mexico', 'Georgia', 'New York', 'Texas', 'Ohio', 'Alabama',
       'Montana', 'Oregon', 'Alaska', 'Louisiana', 'Washington',
       'Minnesota', 'Maine', 'Maryland', 'Illinois', 'California',
       'New Jersey', 'Wisconsin', 'North Carolina', 'Michigan',
       'Pennsylvania', 'Colorado', 'West Virginia', 'Wyoming', 'Utah',
       'South Carolina', 'Iowa', 'Vermont', 'Puerto Rico', 'South Dakota',
       'Tennessee', 'Missouri', 'Massachusetts', 'North Dakota', 'Idaho',
       'Florida', 'Nevada', 'Hawaii',
       'U.S. Pacific Trust Territories and Possessions', 'Kentucky', nan,
       'Arkansas', 'Nebraska', 'Mississippi', 'Indiana', 'New Hampshire',
       'Virginia', 'Connecticut', 'Arizona', 'Delaware', 'Rhode Island',
       'U.S. Virgin Islands', 'Fakestate'], dtype=object)

In [ ]:
## Korzystając z operacji set, znajdź różnicę między stanami początkowymi w naszej ramce danych a naszą predefiniowaną listą
inconsistent_categories = set(flights_df["ORIGINSTATENAME"]).difference(states)
inconsistent_categories

{'Fakestate'}

In [ ]:
# Metoda .isin zwraca wszystkie wiersze z ramki danych, w których został spełniony p[rzekazany warunek]
inconsistent_rows = flights_df["ORIGINSTATENAME"].isin(inconsistent_categories)
flights_df[inconsistent_rows]

TRANSACTIONID  FLIGHTDATE AIRLINECODE               AIRLINENAME  \
1191793      127265600  20130106.0          DL  Delta Air Lines Inc.: DL   

        TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
1191793  N385DN       1033               ATL   

                                           ORIGAIRPORTNAME ORIGINCITYNAME  \
1191793  AtlantaGA: Hartsfield-Jackson Atlanta Internat...        Atlanta   

        ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
1191793          GA       Fakestate             BHM   

                                           DESTAIRPORTNAME DESTCITYNAME  \
1191793  BirminghamAL: Birmingham-Shuttlesworth Interna...   Birmingham   

        DESTSTATE DESTSTATENAME  CRSDEPTIME  DEPTIME  DEPDELAY  TAXIOUT  \
1191793        AL       Alabama        1840   1900.0      20.0     32.0   

         WHEELSOFF  WHEELSON  TAXIIN  CRSARRTIME  ARRTIME  ARRDELAY  \
1191793     1932.0    1859.0     4.0      1838.0   1903.0      25.0   

         CRSELAPSEDTIME  ACTUALELAPSEDTIME CANCELLED DIVERTED  DISTANCE  \
1191793            58.0               63.0     False        F       134   

         AIRLINECODE_ASINT  
1191793                  6

In [ ]:
# Sprytna sztuczka, której możemy użyć do usunięcia wierszy
# Jak myślisz, co oznacza ~?
flights_df = flights_df[~inconsistent_rows]

In [ ]:
set(flights_df['ORIGINSTATENAME'])

### Zmiana mapowania kategorii

To, co widzieliśmy powyżej, to dane, których nie było w predefiniowanym zestawie kategorii. Jednak możemy również natknąć się na inny rodzaj problemów z danymi kategorycznymi, które lepiej rozwiązywać poprzez ponowne mapowanie kategorii niż usuwanie danych. Odpowiednie miejsca do wykonania tego ponownego mapowania to:
1. **Niespójność wartości**: „żonaty”, „niezamężna”, „stanu wolnego”, „ożeniony” <br>
 1. Uważaj też na końcowe białe znaki!
2. **Konwertowanie danych na kategorie lub zbyt wiele kategorii**: Załóżmy, że w naszej ramce danych mamy kolumnę dochodu gospodarstwa domowego.
 1. Możemy zmienić ten typ danych na kategoryczny, grupując dochody (np. `0 - 20k`, `20k - 40k`, `40k - 60k`, `60k +` itd.).
 2. Możemy również zredukować to dalej do `low_class`, `middle_class`, `upper_class`
 
Zajmijmy się nimi w kolejności. W naszej ramce danych lotów kolumny „CANCELLED” i „DIVERTED” przyjmują niespójne wartości. Być może najbezpieczniejszą opcją jest uruchomienie `.value_counts()` na jednej z tych kolumn (`.value_counts()` działa na danych typu `Series`)

In [ ]:
flights_df["CANCELLED"].value_counts()

False    637287
0        347545
F        178357
True      16359
1          8160
T          4084
Name: CANCELLED, dtype: int64

Świetnie! Widzimy więc, że nasze Fałszywe wartości mogą przyjąć jedną z trzech wartości, a wartości Prawdy też są podobne. Możemy arbitralnie wykorzystać te, których chcemy użyć, posuwając się do przodu. Dla jednoznaczności wybierzmy odpowiednio False i True.

In [ ]:
## Skorzystaj z metody replace: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.replace.html
## Aby zamienić 0, F, 1 and T na odpowiadające wartości w kolumnie CANCELLED
flights_df["CANCELLED"].replace(["0", "F"], "False", inplace=True)
flights_df["CANCELLED"].replace(["1", "T"], "True", inplace=True)
flights_df["CANCELLED"].value_counts()

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


False    1163189
True       28603
Name: CANCELLED, dtype: int64

In [ ]:
flights_df["DIVERTED"].value_counts()

F        426570
False    407670
0        354906
T           966
True        881
1           799
Name: DIVERTED, dtype: int64

In [ ]:
# Możemy alternatywnie użyć słownika, aby "zredukować" nasze kategorie.
mapping = {"F": "False", "0": "False", "1": "True", "T": "True"}
flights_df["DIVERTED"] = flights_df["DIVERTED"].replace(mapping)
flights_df["DIVERTED"].value_counts()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


False    1189146
True        2646
Name: DIVERTED, dtype: int64

Jak wspomniano wcześniej, inną sytuacją, w której możemy chcieć ponownie przyporządkować kategorie, jest zmniejszenie liczby wartości w kolumnie. W naszym przypadku załóżmy, że firma lotnicza chciałaby sklasyfikować loty na podstawie długości tras. Tak więc wszystko między 0 a 1000 mil jest `short`, między 1000 a 2500 to `medium`, a 2500+ to `long`.

Możemy użyć metody [`.cut`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html) aby zaklasyfikować dane. Musimy podać 3 argumenty:
1. `Series`, którą chcemy sklasyfikować
2. Bins - liczbę przedziałów
3. Labels - etykiety do przypisania przydziałom

In [ ]:
import numpy as np

bins = [0, 1000, 2500, np.inf]
labels = ["short", "medium", "long"]
flights_df["DISTANCE_CATEGORY"] = pd.cut(flights_df["DISTANCE"], bins=bins, labels=labels)

flights_df[["DISTANCE", "DISTANCE_CATEGORY"]]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


DISTANCE DISTANCE_CATEGORY
0             580             short
1             744             short
2             718             short
3             487             short
4             744             short
...           ...               ...
1191800       721             short
1191801       731             short
1191802       731             short
1191803       508             short
1191804       306             short

[1191792 rows x 2 columns]

In [ ]:
flights_df[flights_df["DISTANCE_CATEGORY"] == "long"]

TRANSACTIONID  FLIGHTDATE AIRLINECODE                 AIRLINENAME  \
3660         120678800  20120103.0          DL    Delta Air Lines Inc.: DL   
5518         117053600  20110506.0          AS    Alaska Airlines Inc.: AS   
5519         117079900  20110523.0          AS    Alaska Airlines Inc.: AS   
5520         120549200  20120102.0          AS    Alaska Airlines Inc.: AS   
5521         120578800  20120122.0          AS    Alaska Airlines Inc.: AS   
...                ...         ...         ...                         ...   
1187039       11875100  19940523.0          AA  American Airlines Inc.: AA   
1187465        8514000  19930925.0          AA  American Airlines Inc.: AA   
1189798      119348600  20110910.0          DL    Delta Air Lines Inc.: DL   
1189810      118919800  20110920.0          AA  American Airlines Inc.: AA   
1190299      130654500  20130927.0          DL    Delta Air Lines Inc.: DL   

        TAILNUM  FLIGHTNUM ORIGINAIRPORTCODE  \
3660     N810NW        837               ATL   
5518     N566AS        897               BLI   
5519     N589AS        897               BLI   
5520     N559AS        897               BLI   
5521     N529AS        897               BLI   
...         ...        ...               ...   
1187039     NaN         73               ORD   
1187465     NaN         73               ORD   
1189798  N554NW       1088               ANC   
1189810  N625AA        278               ANC   
1190299  N821NW        837               ATL   

                                           ORIGAIRPORTNAME ORIGINCITYNAME  \
3660     AtlantaGA: Hartsfield-Jackson Atlanta Internat...        Atlanta   
5518                BellinghamWA: Bellingham International     Bellingham   
5519                BellinghamWA: Bellingham International     Bellingham   
5520                BellinghamWA: Bellingham International     Bellingham   
5521                BellinghamWA: Bellingham International     Bellingham   
...                                                    ...            ...   
1187039            ChicagoIL: Chicago O'Hare International        Chicago   
1187465            ChicagoIL: Chicago O'Hare International        Chicago   
1189798   AnchorageAK: Ted Stevens Anchorage International      Anchorage   
1189810   AnchorageAK: Ted Stevens Anchorage International      Anchorage   
1190299  AtlantaGA: Hartsfield-Jackson Atlanta Internat...        Atlanta   

        ORIGINSTATE ORIGINSTATENAME DESTAIRPORTCODE  \
3660             GA         Georgia             HNL   
5518             WA      Washington             HNL   
5519             WA      Washington             HNL   
5520             WA      Washington             HNL   
5521             WA      Washington             HNL   
...             ...             ...             ...   
1187039          IL        Illinois             HNL   
1187465          IL        Illinois             HNL   
1189798          AK          Alaska             MSP   
1189810          AK          Alaska             DFW   
1190299          GA         Georgia             HNL   

                                           DESTAIRPORTNAME       DESTCITYNAME  \
3660                    HonoluluHI: Honolulu International           Honolulu   
5518                    HonoluluHI: Honolulu International           Honolulu   
5519                    HonoluluHI: Honolulu International           Honolulu   
5520                    HonoluluHI: Honolulu International           Honolulu   
5521                    HonoluluHI: Honolulu International           Honolulu   
...                                                    ...                ...   
1187039                 HonoluluHI: Honolulu International           Honolulu   
1187465                 HonoluluHI: Honolulu International           Honolulu   
1189798   MinneapolisMN: Minneapolis-St Paul International        Minneapolis   
1189810  Dallas/Fort WorthTX: Dallas/Fort Worth Interna...  Dallas/Fort Worth   
1190299           

### Radzenie sobie z danymi datoczasowymi

Jednym z typowych problemów, z którymi się spotkasz, jest zajmowanie się datami i godzinami. Czemu? Ponieważ istnieje wiele sposobów formatowania daty, na przykład `DD/MM/RRRR`, `MM/DD/RR`, `X. MIESIĄC ROK` itp. W powyżej ramce danych nasze daty są w rzeczywistości sformatowane jako jedna numer. Pandas dostarcza nam przydatnego pomocnika do konstruowania dat i godzin — to znaczy metodę [`.to_datetime()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html ).

Zanim się tym zajmiemy, warto szybko przedstawić, w jaki sposób daty są zwykle przechowywane na komputerach. Zazwyczaj daty są obliczane na podstawie liczby sekund, które upłynęły od **1 stycznia 1970**. Gdy chcemy, znaleźć różnicę w czasie między 3.02.2013 16:00 a 21.01.2013 09:00, program wykonuje swoje operacje na **czasie Epoka/Unix/POSIX** dla tych wartości, a następnie możemy coś zakodować, aby otrzymać wartość z powrotem w wybranym przez nas formacie (np. 13 dni, 7 godzin). Posługując się liczbami:

- **3/2/2013 16:00** = 1,359,907,200
- **21/1/2013 09:00** = 1,358,758,800

Różnica w datach = 1,359,907,200 - 1,358,758,800 = 1,148,400 sekund

`format(1148400) = 13 dni, 7 godzin`

Przykłady formatowania dat:
<table>
    <tr>
        <td><b>Date</b></td>
        <td><b>Datetime format</b></td>
    </tr>
    <tr>
        <td>15th June 2020</td>
        <td>%c</td>
    </tr>
    <tr>
        <td>15/06/2020</td>
        <td>%d/%m/%Y</td>
    </tr>
    <tr>
        <td>06-15-2020</td>
        <td>%m-%d-%Y</td>
    </tr>
</table>


In [ ]:
pd.to_datetime(flights_df["FLIGHTDATE"])

0         1970-01-01 00:00:00.020020101
1         1970-01-01 00:00:00.020020101
2         1970-01-01 00:00:00.020020101
3         1970-01-01 00:00:00.020020101
4         1970-01-01 00:00:00.020020103
                       ...             
1191800   1970-01-01 00:00:00.020130106
1191801   1970-01-01 00:00:00.020130106
1191802   1970-01-01 00:00:00.020130106
1191803   1970-01-01 00:00:00.020130106
1191804   1970-01-01 00:00:00.020130106
Name: FLIGHTDATE, Length: 1191792, dtype: datetime64[ns]

To dlaczego teraz wszystkie daty to 1970-01-01?

A to dlatego, że daty są wewnętrznie przechowywane jako sekundy (liczby). Nasza kolumna `FLIGHTDATE` również wyświetla daty lotów w postaci liczb. Tak więc, kiedy uruchamiamy metodę `.to_datetime()`, wszystkie nasze daty są interpretowane jako czas POSIX.

Jednym z prostych rozwiązań, aby to naprawić, jest jawne określenie naszego formatu daty i godziny. Biorąc pod uwagę powyższe przykłady, jak myślisz, jaki będzie format daty?
One simple solution we can do to fix that is to explicitly specify our datetime?

In [ ]:
## Przypisywanie formatu
date_format = "%Y%m%d"
flights_df["FLIGHTDATE"] = pd.to_datetime(flights_df["FLIGHTDATE"], format=date_format)
flights_df["FLIGHTDATE"]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0         2002-01-01
1         2002-01-01
2         2002-01-01
3         2002-01-01
4         2002-01-03
             ...    
1191800   2013-01-06
1191801   2013-01-06
1191802   2013-01-06
1191803   2013-01-06
1191804   2013-01-06
Name: FLIGHTDATE, Length: 1191792, dtype: datetime64[ns]

To rozwiązanie było dość specyficzne dla problemu, który mieliśmy pod ręką. Ale w prawdziwym świecie często można napotkać mieszane formaty dat w jednej ramce danych. Na przykład:

<table>
    <tr>
        <td><b>Name</b></td>
        <td><b>Date of Birth</b></td>
        <td><b>Age</b></td>
    </tr>
    <tr>
        <td>John</td>
        <td>01/07/1995</td>
        <td>25</td>
    </tr>
    <tr>
        <td>Jane</td>
        <td>20-04-1992</td>
        <td>28</td>
    </tr>
    <tr>
        <td>Mark</td>
        <td>3rd January 1990</td>
        <td>30</td>
    </tr>
    </table>


`.to_datetime()` ponownie przychodzi tutaj na ratunek! W poprzedniej komórce kodu wyraźnie ustawiliśmy format daty (ze względu na nietypowy charakter sposobu przechowywania tej daty w ramce danych) - ale bardziej ogólnie możemy użyć `.to_datetime()`, aby automatycznie wywnioskować format każdej daty z osobna.

```python
# errors='coerce' means we'll return NA rows for invalid dates
df["DATE"] = pd.to_datetime(df["DATE"], infer_datetime_format=True, errors='coerce') 
```

## Walidacja krzyżowa

Co to znaczy sprawdzić integralność naszych danych? Zasadniczo musimy mieć świadomość, że kolumna danych, które widzimy, jest spójna w oparciu o inne kolumny danych. Właśnie to sprawdza się w **walidacji krzyżowej**. Zanim rozszerzę niektóre walidacje krzyżowe w tym zbiorze danych, przedstawię nieco bardziej trywialny przykład, aby pokazać, gdzie niewykonanie takich kontroli może zniekształcić analizę:

Poniższa fikcyjna tabela pokazuje wpisy niektórych posiadaczy kredytów studenckich na studia licencjackich (UG) i podyplomowych (PG). Zestaw danych składa się z imienia i nazwiska kredytobiorcy, daty urodzenia (DOB), obecnego wieku (lub wieku zmarłego, jeśli dotyczy, niezależnie od tego, czy zmarli, czy też nie), kwoty ich kredytu UG i PG oraz całkowite kwoty, którą są winni – co powinno być sumą dwóch poprzednich pól. W poniższej tabeli zaznaczono wątpliwe pola kursywą.

<table>
    <tr>
        <td><b>Name</b></td>
        <td><b>D.O.B</b></td>
        <td><b>Age</b></td>
        <td><b>Deceased</b></td>
        <td><b>U.G Loan (£)</b></td>
        <td><b>P.G Loan (£)</b></td>
        <td><b>Total Loan (£)</b></td>
    </tr>
    <tr>
        <td>Idaline</td>
        <td>1971-04-27</td>
        <td>49</td>
        <td>F</td>
        <td>24100</td>
        <td>11900</td>
        <td>36000</td>
    </tr>
    <tr>
        <td>Freddie</td>
        <td>1962-12-27</td>
        <td>57</td>
        <td>F</td>
        <td>26600</td>
        <td>12600</td>
        <td>39200</td>
    </tr>
    <tr>
        <td>Debee</td>
        <td>1970-11-19</td>
        <td>49</td>
        <td>F</td>
        <td>32400</td>
        <td>97000</td>
        <td><i>42100</i></td>
    </tr>
    <tr>
        <td>Joyann</td>
        <td>1957-01-24</td>
        <td><i>41</i></td>
        <td>T</td>
        <td>24400</td>
        <td>11500</td>
        <td>35900</td>
    </tr>
    <tr>
        <td>Ajay</td>
        <td>1960-05-12</td>
        <td><i>50</i></td>
        <td>F</td>
        <td>25500</td>
        <td>18800</td>
        <td>44300</td>
    </tr>
    <tr>
        <td>Emelia</td>
        <td>1957-11-23</td>
        <td><i>57</i></td>
        <td>T</td>
        <td>34000</td>
        <td>17500</td>
        <td><i>0</i></td>
    </tr>
            
</table>
            
            ajay, emelia, joyann

In [ ]:
html_table = """
<table>
    <tr>
        <td><b>Name</b></td>
        <td><b>D.O.B</b></td>
        <td><b>Age</b></td>
        <td><b>Deceased</b></td>
        <td><b>U.G Loan (£)</b></td>
        <td><b>P.G Loan (£)</b></td>
        <td><b>Total Loan (£)</b></td>
    </tr>
    <tr>
        <td>Idaline</td>
        <td>19710427</td>
        <td>50</td>
        <td>F</td>
        <td>24100</td>
        <td>11900</td>
        <td>36000</td>
    </tr>
    <tr>
        <td>Freddie</td>
        <td>19621227</td>
        <td>58</td>
        <td>F</td>
        <td>26600</td>
        <td>12600</td>
        <td>39200</td>
    </tr>
    <tr>
        <td>Debee</td>
        <td>19701119</td>
        <td>49</td>
        <td>F</td>
        <td>32400</td>
        <td>97000</td>
        <td><i>42100</i></td>
    </tr>
    <tr>
        <td>Joyann</td>
        <td>19570124</td>
        <td><i>41</i></td>
        <td>T</td>
        <td>24400</td>
        <td>11500</td>
        <td>35900</td>
    </tr>
    <tr>
        <td>Ajay</td>
        <td>19600512</td>
        <td><i>50</i></td>
        <td>F</td>
        <td>25500</td>
        <td>18800</td>
        <td>44300</td>
    </tr>
    <tr>
        <td>Emelia</td>
        <td>19571123</td>
        <td><i>57</i></td>
        <td>T</td>
        <td>34000</td>
        <td>17500</td>
        <td><i>0</i></td>
    </tr>
            
</table>
"""

html_df = pd.read_html(html_table, header=0)[0]
html_df

Name     D.O.B  Age Deceased  U.G Loan (£)  P.G Loan (£)  Total Loan (£)
0  Idaline  19710427   50        F         24100         11900           36000
1  Freddie  19621227   58        F         26600         12600           39200
2    Debee  19701119   49        F         32400         97000           42100
3   Joyann  19570124   41        T         24400         11500           35900
4     Ajay  19600512   50        F         25500         18800           44300
5   Emelia  19571123   57        T         34000         17500               0

Najpierw popracujmy nad zmienną **Age**. Zgodnie z naszą dokumentacją danych, wiek w komórce powinien odzwierciedlać aktualny wiek kredytobiorców. Wyjątkiem jest sytuacja, gdy kredytobiorca nie żyje, w którym to przypadku wiek powinien zawierać wiek kredytobiorcy w momencie jego śmierci. Najpierw ustalmy, które wiersze łamią ten warunek.

In [ ]:
# najpierw zmieńmy nazwy niektórych kolumn
html_df.columns = ["name", "dob", "age", "deceased", "ug_loan", "pg_loan", "total_loan"]
html_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        6 non-null      object
 1   dob         6 non-null      int64 
 2   age         6 non-null      int64 
 3   deceased    6 non-null      object
 4   ug_loan     6 non-null      int64 
 5   pg_loan     6 non-null      int64 
 6   total_loan  6 non-null      int64 
dtypes: int64(5), object(2)
memory usage: 464.0+ bytes


In [ ]:
## Zamie nmy 'dob' na obiekt date
html_df["dob"] = pd.to_datetime(html_df["dob"], format="%Y%m%d")
html_df

name        dob  age deceased  ug_loan  pg_loan  total_loan
0  Idaline 1971-04-27   50        F    24100    11900       36000
1  Freddie 1962-12-27   58        F    26600    12600       39200
2    Debee 1970-11-19   49        F    32400    97000       42100
3   Joyann 1957-01-24   41        T    24400    11500       35900
4     Ajay 1960-05-12   50        F    25500    18800       44300
5   Emelia 1957-11-23   57        T    34000    17500           0

In [ ]:
# Stwórz nową kolumnę „now_date” wypełnioną aktualną datą i godziną
html_df["now_date"] = pd.Timestamp(pd.datetime.now())

## Oblicz różnicę między 'dob' i 'now_date' i zwróć wartość jako lata
now_date_dob_difference = html_df["now_date"] - html_df["dob"]
now_date_dob_difference

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  


0   18605 days 18:14:47.414006
1   21648 days 18:14:47.414006
2   18764 days 18:14:47.414006
3   23811 days 18:14:47.414006
4   22607 days 18:14:47.414006
5   23508 days 18:14:47.414006
dtype: timedelta64[ns]

In [ ]:
# Ta linia zmienia obiekty timedate na rok zmiennoprzecinkowy, który następnie konwertujemy na int
now_date_dob_difference = (now_date_dob_difference / np.timedelta64(1, 'Y')).astype("int64")
now_date_dob_difference

0    50
1    59
2    51
3    65
4    61
5    64
dtype: int64

In [ ]:
# Na oko możemy zobaczyć, który wiek nie pasuje do ramki danych, którą pokazaliśmy wcześniej.
# Jednak ogólnie zakodujmy to za pomocą logiki pandas.
## Zwróć wiersze, w których „now_date_dob_difference” różni się od zmiennej wieku ramki danych
html_df[html_df["age"] != now_date_dob_difference]

name        dob  age deceased  ug_loan  pg_loan  total_loan  \
1  Freddie 1962-12-27   58        F    26600    12600       39200   
2    Debee 1970-11-19   49        F    32400    97000       42100   
3   Joyann 1957-01-24   41        T    24400    11500       35900   
4     Ajay 1960-05-12   50        F    25500    18800       44300   
5   Emelia 1957-11-23   57        T    34000    17500           0   

                    now_date  
1 2022-04-04 18:14:47.414006  
2 2022-04-04 18:14:47.414006  
3 2022-04-04 18:14:47.414006  
4 2022-04-04 18:14:47.414006  
5 2022-04-04 18:14:47.414006

Przyjrzyjmy się, dlaczego powyższe komórki zostały zwrócone. Jak wspomniano wcześniej, jeśli pożyczkobiorca nie żyje, jego wiek powinien to odzwierciedlać. Oznacza to, że wiek Joyann i Emelii jest rzeczywiście prawidłowy. Używając logiki, odfiltrujmy te wiersze, aby zwrócić tylko te wiersze, które mają matematycznie niepoprawny wiek.

In [ ]:
## Odfiltruj odpowiednich pożyczkodawców za pomocą logiki (wskazówka: &)
incorrect_age_rows = html_df[(html_df["age"] != now_date_dob_difference) & (html_df["deceased"] == "F")]
incorrect_age_rows

name        dob  age deceased  ug_loan  pg_loan  total_loan  \
1  Freddie 1962-12-27   58        F    26600    12600       39200   
2    Debee 1970-11-19   49        F    32400    97000       42100   
4     Ajay 1960-05-12   50        F    25500    18800       44300   

                    now_date  
1 2022-04-04 18:14:47.414006  
2 2022-04-04 18:14:47.414006  
4 2022-04-04 18:14:47.414006

In [ ]:
## Zaktualizuj ramkę danych invalid_age_rows o poprawiony wiek
incorrect_age_rows["age"] = ((incorrect_age_rows["now_date"] - incorrect_age_rows["dob"]) / np.timedelta64(1, 'Y')).astype("int64")
incorrect_age_rows

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


name        dob  age deceased  ug_loan  pg_loan  total_loan  \
1  Freddie 1962-12-27   59        F    26600    12600       39200   
2    Debee 1970-11-19   51        F    32400    97000       42100   
4     Ajay 1960-05-12   61        F    25500    18800       44300   

                    now_date  
1 2022-04-04 18:14:47.414006  
2 2022-04-04 18:14:47.414006  
4 2022-04-04 18:14:47.414006

In [ ]:
## Teraz zaktualizuj odpowiednie wpisy html_df o kolumnę wieku z ramki danych invalid_age_rows
html_df.update(incorrect_age_rows["age"])
html_df

name        dob   age deceased  ug_loan  pg_loan  total_loan  \
0  Idaline 1971-04-27  50.0        F    24100    11900       36000   
1  Freddie 1962-12-27  59.0        F    26600    12600       39200   
2    Debee 1970-11-19  51.0        F    32400    97000       42100   
3   Joyann 1957-01-24  41.0        T    24400    11500       35900   
4     Ajay 1960-05-12  61.0        F    25500    18800       44300   
5   Emelia 1957-11-23  57.0        T    34000    17500           0   

                    now_date  
0 2022-04-04 18:14:47.414006  
1 2022-04-04 18:14:47.414006  
2 2022-04-04 18:14:47.414006  
3 2022-04-04 18:14:47.414006  
4 2022-04-04 18:14:47.414006  
5 2022-04-04 18:14:47.414006

In [ ]:
## Konwertuj wiek z powrotem na int
html_df["age"] = html_df["age"].astype("int64")
## Usuń kolumnę now_date
html_df.drop("now_date", axis=1, inplace=True)
html_df

name        dob  age deceased  ug_loan  pg_loan  total_loan
0  Idaline 1971-04-27   50        F    24100    11900       36000
1  Freddie 1962-12-27   59        F    26600    12600       39200
2    Debee 1970-11-19   51        F    32400    97000       42100
3   Joyann 1957-01-24   41        T    24400    11500       35900
4     Ajay 1960-05-12   61        F    25500    18800       44300
5   Emelia 1957-11-23   57        T    34000    17500           0

Popracujmy teraz nad kwotami pożyczki. Zwróć wszystkie kolumny, w których `ug_loan` + `pg_loan` nie są równe `total_loan`

In [ ]:
## Podzbiór `ug_loan` i `pg_loan` z naszej ramki danych, a następnie sumowanie wzdłuż osi kolumny
sum_loans = html_df[["ug_loan", "pg_loan"]].sum(axis=1)
sum_loans

0     36000
1     39200
2    129400
3     35900
4     44300
5     51500
dtype: int64

In [ ]:
## Zwróć wiersze, które mają nieprawidłowe wartości sum
incorrect_loan_rows = html_df[html_df["total_loan"] != sum_loans]
incorrect_loan_rows

name        dob  age deceased  ug_loan  pg_loan  total_loan
2   Debee 1970-11-19   51        F    32400    97000       42100
5  Emelia 1957-11-23   57        T    34000    17500           0

### Jak postępować z polami, które nie przeszły walidacji?

Tutaj widzimy dwa wiersze, które nie zawierają poprawnych całkowitych kwot kredytu. Analizując każdy z nich z osobna, widzimy, że dane w pierwszym wierszu najprawdopodobniej zawierały niepoprawną wartość `pg_loan` (97 000 GBP w przypadku pożyczki podyplomowej). W drugim, z jakiegoś powodu wartość `total_loan` nie została obliczona. Naiwną strategią może być nadpisanie całkowitych kwot pożyczki sumą `ug_loan` i `pg_loan`. To naprawia typy błędów, w których zwracany jest drugi wiersz. Jednak może istnieć podstawowy problem z powodu pierwszego rzędu. Jeśli zsumujemy tutaj `ug_loan` i `pg_loan`, utworzymy **obserwację odstającą**. W prawdziwym zbiorze danych mogą wystąpić bardzo realne zagrożenia, takie jak te, które mogą narazić na szwank integralność danych – takie problemy mogą łatwo wymykać się spod kontroli, więc upewnij się, że poświęcisz czas na przemyślenie, w jaki sposób Twoje działania wpłyną na Twój dane.

Jak wspomniano wcześniej, niektóre aspekty nauki o danych są sztuką – ale jakąkolwiek decyzję heurystyczną podejmiemy, musimy znaleźć dla niej mocne uzasadnienie. W tym konkretnym przypadku zamierzam usunąć wiersze z niepoprawnym `total_loan`, ponieważ ten błąd prawdopodobnie wystąpił z powodu błędu we wprowadzaniu danych przez człowieka. Wiersze z `total_loan = 0` prawdopodobnie wystąpiły z powodu jakiegoś systematycznego błędu - być może z innej bazy danych, w której nie podano sumy total_loan. Biorąc pod uwagę inne weryfikacje, jednym z rozwiązań, które moglibyśmy wybrać, jest zsumowanie dwóch kolumn.

In [ ]:
## Zidentyfikuj wiersze, w których total_loan NIE jest równe 0, ale jest niepoprawne
incorrect_loan_but_not_zero_rows = html_df[(html_df["total_loan"] != sum_loans) & (html_df["total_loan"] != 0)]
print(incorrect_loan_but_not_zero_rows)

## Usuń te wiersze
html_df = html_df.drop(incorrect_loan_but_not_zero_rows.index)

html_df

    name        dob  age deceased  ug_loan  pg_loan  total_loan
2  Debee 1970-11-19   51        F    32400    97000       42100


name        dob  age deceased  ug_loan  pg_loan  total_loan
0  Idaline 1971-04-27   50        F    24100    11900       36000
1  Freddie 1962-12-27   59        F    26600    12600       39200
3   Joyann 1957-01-24   41        T    24400    11500       35900
4     Ajay 1960-05-12   61        F    25500    18800       44300
5   Emelia 1957-11-23   57        T    34000    17500           0

In [ ]:
html_df[(html_df["total_loan"] == 0)]


name        dob  age deceased  ug_loan  pg_loan  total_loan
5  Emelia 1957-11-23   57        T    34000    17500           0

In [ ]:
html_df.drop(index=5, inplace=True)

In [ ]:
# Zakładając, że jesteśmy zadowoleni ze wszystkich innych wpisów w naszych pożyczkobiorcach, możemy bezpośrednio obliczyć i nadpisać total_loan w naszej ramce danych
## Zastąp total_loan sumą ug_loan i pg_loan
html_df["total_loan"] = html_df["ug_loan"] + html_df["pg_loan"]
html_df

## Praca z danymi tekstowymi i typu string

Dane tekstowe są oczywiście niezwykle powszechnym rodzajem danych i mogą przybierać różne formy – od tekstu nieustrukturyzowanego po e-maile, nazwiska, numery telefonów itp. Istnieje wiele rodzajów problemów, które możemy napotkać w przypadku danych tekstowych:
- Niespójność danych (np. +86 195 448 8582 vs 0086-195-448-8582)
- Naruszenia tekstu (np. niedozwolone znaki, błędy w polach wejściowych, literówki w tekście)
- Literówki „strukturalne” (np. +86.1954.48858.2)

W przykładowej tabeli poniżej widzimy listę osób wraz z ich imionami i numerami telefonów. Jak widać – najprawdopodobniej ze względu na wolne pola tekstowe, nazwiska i numery telefonów zostały wprowadzone w różnych formatach. Naszym zadaniem jest standaryzacja tych pól, aby były spójne w całej ramce danych:

<table>
    <tr>
        <td><b>Name</b></td>
        <td><b>Phone Number</b></td>
    </tr>
    <tr>
        <td>Dr Darci Abela</td>
        <td>+86-185-338-1819</td>
    </tr>
    <tr>
        <td>Mr Patten St. Queintain</td>
        <td>00865872411917</td>
    </tr>
    <tr>
        <td>mr conant burden</td>
        <td>0086-289-702-0948</td>
    </tr>
    <tr>
        <td>miss marcia Dutnell</td>
        <td>0668</td>
    </tr>
    <tr>
        <td>dr Greggory lurner</td>
        <td>+31 778 813 8432</td>
    </tr>
    <tr>
        <td>MS Doe Beavan</td>
        <td>+420-731-276-7633</td>
    </tr>
    <tr>
        <td>Tamarah Delgado</td>
        <td>+868431029051</td>
    </tr>
    <tr>
        <td>Miss Arlee daborne</td>
        <td>+33-307-220-2746</td>
    </tr>
    <tr>
        <td>Ly b. Grima</td>
        <td>+238-863-946-4232</td>
    </tr>
</table>

Użyjemy małego sztucznie stworzonego zbioru danych w csv w tym celu.

In [ ]:
# np = names_phones
np_df = pd.read_csv(f"{path}mock_names_phones.csv", header=0, index_col=0)
np_df

name    phone number
0   Dr rafaello Vlasenko    215-183-0246
1    miss Dido Rosbrough    913 520 0662
2   MISS Drusie Merriday            0355
3     MRS Jillayne Kiloh   0087-407-5997
4   Rev VALERIA SHEVLANE    916-308-0837
..                   ...             ...
95      REV gun Ornillos   0036 831 9242
96      Rev Peg Splevins    212-524-1998
97     Mr Rosalind Beyer  + 447 730 3458
98    Dr Wenonah O' Hern   0036-477-2705
99     Ms PATTIE SOPPETT  00622-354-4677

[100 rows x 2 columns]

Ok - dla tego dataframe są 4 zadania:
1. Utwórz kolumnę title, która zawiera tytuł każdej osoby (np. Pani, Panna itp.). Ta kolumna powinna być znormalizowana i kategoryczna
2. Podziel kolumnę name na kolumnę first name i last name. Obie kolumny powinny mieć pierwszą literę imienia z wielkiej litery
3. Usuń wiersz `name`
4. Standaryzuj numery telefonów w formacie `00XXXXXXXXX`. To znaczy - dwa zera poprzedzone resztą rzeczywistej liczby

Zajmijmy się nimi w kolejności

In [ ]:
# Najpierw chcemy utworzyć nową kolumnę tytułową, która przyjmuje tytuły grzecznościowe w kolumnie imienia
# Aby to uzyskać, musimy podzielić name po białym znaku i wziąć pierwszy element z listy podzielonej
example_string = "this string will be split"
print(example_string.split())
print(example_string.split()[0])

['this', 'string', 'will', 'be', 'split']
this


In [ ]:
# Aby wykonać operacje na stringach naa kolumnach stringów w pandas, musimy poprzedzić naszą funkcję ciągów znakiem „.str”
np_df["name"].str.split()


0     [Dr, rafaello, Vlasenko]
1      [miss, Dido, Rosbrough]
2     [MISS, Drusie, Merriday]
3       [MRS, Jillayne, Kiloh]
4     [Rev, VALERIA, SHEVLANE]
                ...           
95        [REV, gun, Ornillos]
96        [Rev, Peg, Splevins]
97       [Mr, Rosalind, Beyer]
98     [Dr, Wenonah, O', Hern]
99       [Ms, PATTIE, SOPPETT]
Name: name, Length: 100, dtype: object

In [ ]:
def capitalize(txt):
  return txt.capitalize()

In [ ]:
## Utwórz i wypełnij kolumnę title.
# To zadanie można rozwiązać na kilka różnych sposobów.
# Zobacz, ile rozwiązań możesz wymyślić
np_df["title"] = np_df["name"].str.split().apply(lambda x: x[0])
np_df

name    phone number title
0   Dr rafaello Vlasenko    215-183-0246    Dr
1    miss Dido Rosbrough    913 520 0662  miss
2   MISS Drusie Merriday            0355  MISS
3     MRS Jillayne Kiloh   0087-407-5997   MRS
4   Rev VALERIA SHEVLANE    916-308-0837   Rev
..                   ...             ...   ...
95      REV gun Ornillos   0036 831 9242   REV
96      Rev Peg Splevins    212-524-1998   Rev
97     Mr Rosalind Beyer  + 447 730 3458    Mr
98    Dr Wenonah O' Hern   0036-477-2705    Dr
99     Ms PATTIE SOPPETT  00622-354-4677    Ms

[100 rows x 3 columns]

In [ ]:
# Chcemy, aby nasz title był ustandaryzowany i kategoryczny.
## Zamień kolumnę na kolumnę kategoryczną i zwróć wszystkie kategorie, które obecnie istnieją w kolumnie
np_df["title"] = np_df["title"].astype("category")
set(np_df["title"])

{'DR',
 'Dr',
 'MISS',
 'MR',
 'MRS',
 'Miss',
 'Mr',
 'Mrs',
 'Ms',
 'REV',
 'Rev',
 'dr',
 'miss',
 'mr',
 'mrs',
 'ms',
 'rev'}

In [ ]:
# Widzimy wiele różnych wariantów. Wybierzmy metodę normalizacji wpisów (np. wielkie litery).
## Standaryzuj kolumnę title
np_df["title"] = np_df["title"].str.upper()
np_df

name    phone number title
0   Dr rafaello Vlasenko    215-183-0246    DR
1    miss Dido Rosbrough    913 520 0662  MISS
2   MISS Drusie Merriday            0355  MISS
3     MRS Jillayne Kiloh   0087-407-5997   MRS
4   Rev VALERIA SHEVLANE    916-308-0837   REV
..                   ...             ...   ...
95      REV gun Ornillos   0036 831 9242   REV
96      Rev Peg Splevins    212-524-1998   REV
97     Mr Rosalind Beyer  + 447 730 3458    MR
98    Dr Wenonah O' Hern   0036-477-2705    DR
99     Ms PATTIE SOPPETT  00622-354-4677    MS

[100 rows x 3 columns]

In [ ]:
## W podobny sposób do powyższego utwórz nową kolumnę na imię i jedną na nazwisko.
# Upewnij się, że dla obu nowych kolumn nazwy są pisane małymi literami, z wyjątkiem pierwszej litery, która jest pisana wielką literą
np_df["first_name"] = np_df["name"].str.split().apply(lambda x: x[1].lower())
np_df["last_name"] = np_df["name"].str.split().apply(lambda x: x[-1].lower())
np_df["first_name"] = np_df["first_name"].str[0].str.upper() + np_df["first_name"].str[1:]
np_df["last_name"] = np_df["last_name"].str[0].str.upper() + np_df["last_name"].str[1:]

np_df

name    phone number title first_name  last_name
0   Dr rafaello Vlasenko    215-183-0246    DR   Rafaello   Vlasenko
1    miss Dido Rosbrough    913 520 0662  MISS       Dido  Rosbrough
2   MISS Drusie Merriday            0355  MISS     Drusie   Merriday
3     MRS Jillayne Kiloh   0087-407-5997   MRS   Jillayne      Kiloh
4   Rev VALERIA SHEVLANE    916-308-0837   REV    Valeria   Shevlane
..                   ...             ...   ...        ...        ...
95      REV gun Ornillos   0036 831 9242   REV        Gun   Ornillos
96      Rev Peg Splevins    212-524-1998   REV        Peg   Splevins
97     Mr Rosalind Beyer  + 447 730 3458    MR   Rosalind      Beyer
98    Dr Wenonah O' Hern   0036-477-2705    DR    Wenonah       Hern
99     Ms PATTIE SOPPETT  00622-354-4677    MS     Pattie    Soppett

[100 rows x 5 columns]

In [ ]:
## Usuń kolumnę name
np_df.drop("name", axis=1, inplace=True)
np_df

phone number title first_name  last_name
0     215-183-0246    DR   Rafaello   Vlasenko
1     913 520 0662  MISS       Dido  Rosbrough
2             0355  MISS     Drusie   Merriday
3    0087-407-5997   MRS   Jillayne      Kiloh
4     916-308-0837   REV    Valeria   Shevlane
..             ...   ...        ...        ...
95   0036 831 9242   REV        Gun   Ornillos
96    212-524-1998   REV        Peg   Splevins
97  + 447 730 3458    MR   Rosalind      Beyer
98   0036-477-2705    DR    Wenonah       Hern
99  00622-354-4677    MS     Pattie    Soppett

[100 rows x 4 columns]

Świetnie! To prowadzi nas do czwartej części zadania - ujednolicenia numeru telefonu i przekonwertowania go na typ danych int. Przypomnij sobie, jak chcemy, aby nasze numery telefonów wyglądały: zacznij od 00, a następnie do reszty numeru.

In [ ]:
# Zwraca wszystkie (unikalne) numery telefonów, dzięki czemu możemy zobaczyć różne typy problemów jakie nas czekają
set(np_df["phone number"])

{' 215-183-0246',
 ' 308-345-9376',
 ' 372 875 9524',
 ' 575-722-2771',
 ' 639 907 8152',
 ' 731-174-0498',
 ' 748 111 3081',
 ' 913 520 0662',
 '+  103-221-6651',
 '+  254-790-3115',
 '+  265-672-9320',
 '+  878-773-9133',
 '+ 116-341-3550',
 '+ 161-935-4043',
 '+ 170-474-6125',
 '+ 307-275-6256',
 '+ 397-504-0587',
 '+ 405-653-2307',
 '+ 426-945-0672',
 '+ 429-936-7430',
 '+ 447 730 3458',
 '+ 469-180-8942',
 '+ 477-531-0695',
 '+ 545-108-6180',
 '+ 580-877-2968',
 '+ 616-211-4527',
 '+ 677-669-1931',
 '+ 695-276-8827',
 '+ 711-239-5519',
 '+ 758-850-2395',
 '+ 790-814-7053',
 '+ 824 616 7624',
 '+ 832 867 6108',
 '+ 843-164-5335',
 '+ 873-735-5162',
 '+ 950-576-6969',
 '+ 984-358-0707',
 '0006-275-2109',
 '0012 717 7388',
 '00207 236 3362',
 '0026-481-0716',
 '0033-814-2371',
 '0036 831 9242',
 '0036-477-2705',
 '0041-913-3917',
 '0044-670-6014',
 '0057-698-8700',
 '00622-354-4677',
 '00675-449-3771',
 '0070-431-3190',
 '00800-106-0189',
 '0087-407-5997',
 '0096-920-9952',
 '0355',


Ok, więc jakie problemy widzisz?
<details>
     <summary><b>> Kliknij tutaj, aby zobaczyć problemy</b></summary>
     <ul>
         <li>Numery zaczynają się różnie – niektóre zaczynają się od `+`, inne od `00`</li>
         <li>Niektóre numery mają spacje między grupami liczb, inne są dzielone. Niektóre numery również nie mają „grup”</li>
         <li>Niektóre numery zaczynają się od spacji, inne od `+ `, inne od `+`.</li>
         <li>Niektóre numery mają tylko cztery liczby</li>
     </ul>
</details>

Istnieje kilka sposobów formatowania tych ciągów do pożądanego wyniku. Tutaj poprowadzę cię przez metodę, w której iterujemy wiersze i stosujemy funkcję, aby ponownie przypisać zmienną. Zacznijmy od stworzenia funkcji pośredniej, która pobiera numer telefonu i manipuluje nim do pożądanego wyniku.

In [ ]:
def standardise_phone_number(phone_number):
    
    ## jeśli pierwszym znakiem jest "+", usuń go.
    if phone_number.startswith("+"):
        phone_number = phone_number.replace("+", "")
    
    ## usuń wszystkie spacje z numeru telefonu
    phone_number = phone_number.replace(" ", "")
    
    ## usuń myślniki z numeru telefonu
    phone_number = phone_number.replace("-", "")
    
    ## jeśli numer nie zaczyna się od 00, dodaj 00 do początku numeru
    if not phone_number.startswith("00"):
        phone_number = "00" + phone_number
    
    ## zwróć numer telefonu
    return phone_number

In [ ]:
# Będziemy iterować po wierszach ramki danych i ponownie przypiszemy wiersz do standardowego wariantu
for index, row in np_df.iterrows():
    
    ## Odwołaj się do naszej funkcji standaryzacji na numer telefonu dla bieżącej pętli
    row["phone number"] = standardise_phone_number(row["phone number"])
    
np_df

phone number title first_name  last_name
0   002151830246    DR   Rafaello   Vlasenko
1   009135200662  MISS       Dido  Rosbrough
2         000355  MISS     Drusie   Merriday
3    00874075997   MRS   Jillayne      Kiloh
4   009163080837   REV    Valeria   Shevlane
..           ...   ...        ...        ...
95   00368319242   REV        Gun   Ornillos
96  002125241998   REV        Peg   Splevins
97  004477303458    MR   Rosalind      Beyer
98   00364772705    DR    Wenonah       Hern
99  006223544677    MS     Pattie    Soppett

[100 rows x 4 columns]

In [ ]:
# W naszej ramce danych nadal znajdują się nieprawidłowe liczby (tj. te, które pierwotnie miały długość 4)
## Zamień wszystkie numery telefonów poniżej 10 cyfr/znaków na pd.NA
# Podpowiedź: będzie potrzebna metoda .loc
np_df.loc[np_df["phone number"].str.len() < 10, "phone number"] = pd.NA
np_df

phone number title first_name  last_name
0   002151830246    DR   Rafaello   Vlasenko
1   009135200662  MISS       Dido  Rosbrough
2           <NA>  MISS     Drusie   Merriday
3    00874075997   MRS   Jillayne      Kiloh
4   009163080837   REV    Valeria   Shevlane
..           ...   ...        ...        ...
95   00368319242   REV        Gun   Ornillos
96  002125241998   REV        Peg   Splevins
97  004477303458    MR   Rosalind      Beyer
98   00364772705    DR    Wenonah       Hern
99  006223544677    MS     Pattie    Soppett

[100 rows x 4 columns]

In [ ]:
# Skoncentrujemy się na brakujących danych w następnej części, ale policzmy liczbę wierszy z NA i usuńmy je
null_phone_numbers = np_df["phone number"].isnull()
print("Number of null phone numbers:", null_phone_numbers.sum())

# Usuń wiersze, które mają puste numery telefonów
np_df = np_df.dropna(subset=["phone number"])
np_df

Bardziej skomplikowane manipulacje ciągami znaków można wykonać za pomocą **wyrażeń regularnych**, znanych również jako [regex](https://docs.python.org/3/howto/regex.html). Nie będziemy tutaj przyglądać się wyrażeniu regularnemu, ale ważne jest, aby wiedzieć o jego mocy. Zasadniczo wyrażenie regularne pozwala nam określić reguły dla ciągów, które chcemy dopasować. Ma bardzo szerokie zastosowanie. Oto kilka przykładów:
- identyfikuj wiadomości e-mail w zakresach tekstu
- sprawdź, czy adres URL ma poprawny format
- wyodrębnij tylko cyfry z ciągu tekstowego

Kiedy natkniesz się na zadania, które wymagają oczyszczenia danych tekstowych, regex jest narzędziem do tego zadania.

## Scalanie danych razem (merging)

Czasami znajdziemy dane w różnych plikach, które musimy połączyć w jedną ramkę danych, zanim będziemy mogli jej użyć. Przykładem tego są [dane IMDB](http://www.imdb.com/interfaces#plain) – gdzie, być może ze względu na rozmiar dostępnych danych, twórcy postanowili podzielić cały zbiór danych na wiele osobnych plików. Przyjrzyjmy się szybko, jak możemy połączyć dane z tych wielu plików.

Źródło: 
IMDb
(http://www.imdb.com).

Pliki do pobrania i scalenia to:
- tytuł.akas.tsv.gz
- tytuł.podstawy.tsv.gz
- tytuł.oceny.tsv.gz

In [ ]:
# URUCHOM TĘ KOMÓRKĘ TYLKO JEŚLI CHCESZ PONOWNIE POBRAĆ CAŁE DANE
import requests
import gzip
import pandas as pd
import pickle
import os

DATA_FOLDER = "DATA" # TODO: najpierw ten folder trzeba utworzyc
DOWNLOAD_URL = "https://datasets.imdbws.com/"
files_to_download = ["title.ratings.tsv.gz", "title.akas.tsv.gz", "title.basics.tsv.gz"]


for file_string in files_to_download:
    # pobierz plik tsv.gz
    # plik jest obiektem request
    print("Getting request for file:", file_string)
    df = pd.read_csv(DOWNLOAD_URL + file_string, sep="\t", compression="gzip")
    df = df.iloc[-10000:]
    print(df.shape)
    pickle.dump(df, open(os.path.join(DATA_FOLDER, file_string.replace(".tsv.gz", ".df")), "wb"))
    print(df.head())


In [ ]:
ratings_df = pickle.load(open(os.path.join(DATA_FOLDER, "title.ratings.df"), "rb"))
films_df = pickle.load(open(os.path.join(DATA_FOLDER, "title.akas.df"), "rb"))
basics_df = pickle.load(open(os.path.join(DATA_FOLDER, "title.basics.df"), "rb"))

### Łączenie wielu ramek danych

Podobnie jak w SQL, chcemy wykonać jakiś rodzaj [`.join()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html) lub [ `.merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html#pandas.DataFrame.merge) przez ramki danych, które obecnie posiadamy. Wymaga to ustawienia unikalnego identyfikatora/klucza podstawowego jako indeksu ramek danych. Pamiętaj o dwóch głównych typach złączeń,  **inner** i **lefyt**. Jak myślisz, które podejście będzie bardziej poprawne w tym przypadku?

In [ ]:
## zmodyfikuj indeksy ratings_df, basics_df i movies_df, aby wszystkie miały wspólny indeks
ratings_df = ratings_df.set_index("tconst")
basics_df = basics_df.set_index("tconst")
films_df = films_df.set_index("titleId")
films_df.head()

In [ ]:
## Wykonaj join lub merge z basics_df i ratings_df za pomocą movies_df
df = films_df.join([basics_df, ratings_df])
df

---